// @plentymarkets/terra-components/app/components/file-browser/model/terra-upload-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("util");
var path_helper_1 = require("../../../helpers/path.helper");
var TerraUploadItem = /** @class */ (function () {
    function TerraUploadItem(file, path, uploadService) {
        this.file = file;
        this.path = path;
        this.uploadService = uploadService;
        this.uploaded = false;
        this.beforeUploadList = [];
        this.onSuccessList = [];
        this.onCancelList = [];
        this.onErrorList = [];
        this.onProgressList = [];
        if (util_1.isNullOrUndefined(file)) {
            this.uploaded = true;
        }
        if (!util_1.isNullOrUndefined(this.path) && this.path.charAt(0) === '/') {
            this.path = this.path.substr(1);
        }
    }
    Object.defineProperty(TerraUploadItem.prototype, "filename", {
        get: function () {
            var filenames = this.file.name.split('.');
            var extname = filenames.pop();
            return this.uploadService.prepareKey(this.file.name, true);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraUploadItem.prototype, "pathname", {
        get: function () {
            var pathname = this.uploadService.prepareKey(path_helper_1.PathHelper.join(this.path, this.filename));
            if (pathname.charAt(0) === '/') {
                pathname = pathname.substr(1);
            }
            return pathname;
        },
        enumerable: true,
        configurable: true
    });
    TerraUploadItem.prototype.beforeUpload = function (callback) {
        this.beforeUploadList.push(callback);
        return this;
    };
    TerraUploadItem.prototype.onSuccess = function (callback) {
        this.onSuccessList.push(callback);
        return this;
    };
    TerraUploadItem.prototype.onError = function (callback) {
        this.onErrorList.push(callback);
        return this;
    };
    TerraUploadItem.prototype.onCancel = function (callback) {
        this.onCancelList.push(callback);
        return this;
    };
    TerraUploadItem.prototype.onProgress = function (callback) {
        this.onProgressList.push(callback);
        return this;
    };
    TerraUploadItem.prototype.cancelUpload = function () {
        this.uploadService.queue.remove(this);
        if (!util_1.isNullOrUndefined(this.xhr)) {
            this.xhr.abort();
        }
    };
    TerraUploadItem.prototype.emit = function (event) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (['beforeUpload',
            'onSuccess',
            'onError',
            'onCancel',
            'onProgress'].indexOf(event) >= 0) {
            this['_' + event].forEach(function (callback) {
                callback.apply(void 0, args);
            });
            if (['onSuccess',
                'onError',
                'onCancel'].indexOf(event) >= 0) {
                this.uploaded = true;
            }
        }
    };
    TerraUploadItem.DONE = new TerraUploadItem(null, null, null);
    return TerraUploadItem;
}());
exports.TerraUploadItem = TerraUploadItem;
