// @plentymarkets/terra-components/app/components/forms/form/model/terra-form-scope.data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("util");
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var TerraFormScope = /** @class */ (function () {
    function TerraFormScope(_data, parent) {
        if (_data === void 0) { _data = {}; }
        this._data = _data;
        this.parent = parent;
        this.onDataChanged = new BehaviorSubject_1.BehaviorSubject({});
    }
    Object.defineProperty(TerraFormScope.prototype, "data", {
        get: function () {
            return this._data;
        },
        set: function (value) {
            this._data = value;
            this.onDataChanged.next(value);
        },
        enumerable: true,
        configurable: true
    });
    TerraFormScope.prototype.evaluate = function (expression) {
        try {
            if (util_1.isNullOrUndefined(this._data)) {
                return (new Function('return ' + expression)).apply(null);
            }
            else {
                var data_1 = this.getEvaluationData();
                var keys = Object.keys(data_1);
                var values = keys.map(function (key) { return data_1[key] || null; });
                return (new (Function.bind.apply(Function, [void 0].concat(keys, ['return ' + expression])))()).apply(null, values);
            }
        }
        catch (e) {
            return null;
        }
    };
    TerraFormScope.prototype.createChildScope = function (data) {
        if (data === void 0) { data = {}; }
        var scope = new TerraFormScope(data, this);
        this.onDataChanged.subscribe(function (value) {
            scope.onDataChanged.next(scope.data);
        });
        return scope;
    };
    TerraFormScope.prototype.getEvaluationData = function () {
        var _this = this;
        var result = {};
        if (!util_1.isNullOrUndefined(this.parent)) {
            result = this.parent.getEvaluationData();
        }
        Object.keys(this._data).forEach(function (key) {
            result[key] = _this._data[key] || null;
        });
        return result;
    };
    return TerraFormScope;
}());
exports.TerraFormScope = TerraFormScope;
