// @plentymarkets/terra-components/app/components/forms/form/model/terra-form-type-map.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _1 = require("../../../../../");
var TerraFormTypeMap = /** @class */ (function () {
    function TerraFormTypeMap() {
        this.checkbox = {
            inputMap: {
                name: 'inputCaption'
            },
            component: _1.TerraCheckboxComponent
        };
        this.date = _1.TerraDatePickerComponent;
        this.file = _1.TerraFileInputComponent;
        this.text = _1.TerraTextInputComponent;
        this.textarea = _1.TerraTextAreaInputComponent;
        this.number = _1.TerraNumberInputComponent;
        this.double = _1.TerraDoubleInputComponent;
        this.select = _1.TerraSelectBoxComponent;
        this.category = _1.TerraCategoryPickerComponent;
        this.color = _1.TerraColorPickerComponent;
        this.checkboxGroup = _1.TerraMultiCheckBoxComponent;
        this.noteEditor = {
            inputMap: {
                name: 'inputHeaderLabel'
            },
            component: _1.TerraNoteEditorComponent
        };
        this.codeEditor = {
            inputMap: {
                name: 'inputHeaderLabel'
            },
            component: _1.TerraCodeEditorComponent
        };
    }
    return TerraFormTypeMap;
}());
exports.TerraFormTypeMap = TerraFormTypeMap;
