// @plentymarkets/terra-components/app/components/layouts/card/terra-card.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraCardComponent = /** @class */ (function () {
    function TerraCardComponent() {
        this.inputIsSelected = false;
        this.showHeader = false;
        this.showFooter = false;
    }
    TerraCardComponent.prototype.ngAfterContentChecked = function () {
        this.showHeader = this.viewChildHeader.nativeElement.children.length > 0;
        this.showFooter = this.viewChildFooter.nativeElement.children.length > 0;
    };
    __decorate([
        core_1.ViewChild('header'),
        __metadata("design:type", core_1.ElementRef)
    ], TerraCardComponent.prototype, "viewChildHeader", void 0);
    __decorate([
        core_1.ViewChild('footer'),
        __metadata("design:type", core_1.ElementRef)
    ], TerraCardComponent.prototype, "viewChildFooter", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraCardComponent.prototype, "inputImagePath", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraCardComponent.prototype, "inputPlaceholderIcon", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraCardComponent.prototype, "inputIsSelected", void 0);
    TerraCardComponent = __decorate([
        core_1.Component({
            selector: 'terra-card',
            styles: [require('./terra-card.component.scss')],
            template: require('./terra-card.component.html')
        }),
        __metadata("design:paramtypes", [])
    ], TerraCardComponent);
    return TerraCardComponent;
}());
exports.TerraCardComponent = TerraCardComponent;
