// @plentymarkets/terra-components/app/components/buttons/file-chooser/terra-file-chooser.component.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var angular_l10n_1 = require("angular-l10n");
var util_1 = require("util");
var terra_file_browser_component_1 = require("../../file-browser/terra-file-browser.component");
var terra_button_component_1 = require("../button/terra-button.component");
var terra_overlay_component_1 = require("../../layouts/overlay/terra-overlay.component");
var terra_frontend_storage_service_1 = require("../../file-browser/terra-frontend-storage.service");
var TerraFileChooserComponent = /** @class */ (function (_super) {
    __extends(TerraFileChooserComponent, _super);
    function TerraFileChooserComponent(translation, frontendStorageService) {
        var _this = _super.call(this) || this;
        _this.translation = translation;
        _this.frontendStorageService = frontendStorageService;
        _this.inputAllowedExtensions = [];
        _this.inputAllowFolders = true;
        _this.outputSelected = new core_1.EventEmitter();
        _this.outputCancelled = new core_1.EventEmitter();
        _this.outputFileBrowserShow = new core_1.EventEmitter();
        _this.outputFileBrowserHide = new core_1.EventEmitter();
        _this.translationPrefix = 'terraFileInput';
        _this.primaryBrowserButtonCaption = '';
        _this.secondaryBrowserButtonCaption = '';
        _this.primaryOverlayButton = {
            icon: 'icon-success',
            caption: _this.inputPrimaryBrowserButtonCaption,
            isDisabled: true,
            clickFunction: function () {
                _this.outputSelected.emit(_this.selectedObject);
                _this.overlay.hideOverlay();
            }
        };
        _this.secondaryOverlayButton = {
            icon: 'icon-close',
            caption: _this.inputSecondaryBrowserButtonCaption,
            isDisabled: false,
            clickFunction: function () {
                _this.outputCancelled.emit();
                _this.overlay.hideOverlay();
            }
        };
        return _this;
    }
    Object.defineProperty(TerraFileChooserComponent.prototype, "inputPrimaryBrowserButtonCaption", {
        get: function () {
            if (!util_1.isNullOrUndefined(this.primaryBrowserButtonCaption) && this.primaryBrowserButtonCaption.length > 0) {
                return this.primaryBrowserButtonCaption;
            }
            return this.translation.translate(this.translationPrefix + '.choose');
        },
        set: function (value) {
            this.primaryBrowserButtonCaption = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraFileChooserComponent.prototype, "inputSecondaryBrowserButtonCaption", {
        get: function () {
            if (!util_1.isNullOrUndefined(this.secondaryBrowserButtonCaption) && this.secondaryBrowserButtonCaption.length > 0) {
                return this.secondaryBrowserButtonCaption;
            }
            return this.translation.translate(this.translationPrefix + '.cancel');
        },
        set: function (value) {
            this.primaryBrowserButtonCaption = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraFileChooserComponent.prototype, "inputStorageServices", {
        get: function () {
            return this.storageServices || [this.frontendStorageService];
        },
        set: function (services) {
            this.storageServices = services;
        },
        enumerable: true,
        configurable: true
    });
    TerraFileChooserComponent.prototype.onClick = function (event) {
        this.outputClicked.emit(event);
        this.overlay.showOverlay();
    };
    TerraFileChooserComponent.prototype.onSelectedObjectChange = function (selectedObject) {
        if (util_1.isNullOrUndefined(selectedObject) || selectedObject.isDirectory) {
            this.primaryOverlayButton.isDisabled = true;
        }
        else {
            this.primaryOverlayButton.isDisabled = false;
            this.selectedObject = selectedObject;
        }
    };
    TerraFileChooserComponent.prototype.onBrowserShow = function () {
        this.outputFileBrowserShow.emit(this.fileBrowser);
    };
    TerraFileChooserComponent.prototype.onBrowserHide = function () {
        this.outputFileBrowserHide.emit(this.fileBrowser);
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], TerraFileChooserComponent.prototype, "inputPrimaryBrowserButtonCaption", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], TerraFileChooserComponent.prototype, "inputSecondaryBrowserButtonCaption", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array)
    ], TerraFileChooserComponent.prototype, "inputAllowedExtensions", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraFileChooserComponent.prototype, "inputAllowFolders", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array),
        __metadata("design:paramtypes", [Array])
    ], TerraFileChooserComponent.prototype, "inputStorageServices", null);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraFileChooserComponent.prototype, "outputSelected", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraFileChooserComponent.prototype, "outputCancelled", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraFileChooserComponent.prototype, "outputFileBrowserShow", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraFileChooserComponent.prototype, "outputFileBrowserHide", void 0);
    __decorate([
        core_1.ViewChild('overlay'),
        __metadata("design:type", terra_overlay_component_1.TerraOverlayComponent)
    ], TerraFileChooserComponent.prototype, "overlay", void 0);
    __decorate([
        core_1.ViewChild('fileBrowser'),
        __metadata("design:type", terra_file_browser_component_1.TerraFileBrowserComponent)
    ], TerraFileChooserComponent.prototype, "fileBrowser", void 0);
    TerraFileChooserComponent = __decorate([
        core_1.Component({
            selector: 'terra-file-chooser',
            template: require('./terra-file-chooser.component.html'),
            styles: [require('./terra-file-chooser.component.scss')]
        }),
        __metadata("design:paramtypes", [angular_l10n_1.TranslationService, terra_frontend_storage_service_1.TerraFrontendStorageService])
    ], TerraFileChooserComponent);
    return TerraFileChooserComponent;
}(terra_button_component_1.TerraButtonComponent));
exports.TerraFileChooserComponent = TerraFileChooserComponent;
