// @plentymarkets/terra-components/app/components/forms/checkbox/example/terra-checkbox.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_checkbox_component_1 = require("../terra-checkbox.component");
var TerraCheckboxComponentExample = /** @class */ (function () {
    function TerraCheckboxComponentExample() {
    }
    TerraCheckboxComponentExample.prototype.ngOnInit = function () {
        this.viewChildIndeterminateCb.isIndeterminate = true;
    };
    TerraCheckboxComponentExample.prototype.getCheckboxIndeterminateState = function () {
        alert('Indeterminate state: ' + this.viewChildIndeterminateCb.isIndeterminate);
    };
    TerraCheckboxComponentExample.prototype.setCheckboxIndeterminateState = function () {
        this.viewChildIndeterminateCb.isIndeterminate = !this.viewChildIndeterminateCb.isIndeterminate;
    };
    TerraCheckboxComponentExample.prototype.setCheckboxValue = function () {
        this.viewChildTestCb.value = !this.viewChildTestCb.value;
    };
    __decorate([
        core_1.ViewChild('viewChildIndeterminateCb'),
        __metadata("design:type", terra_checkbox_component_1.TerraCheckboxComponent)
    ], TerraCheckboxComponentExample.prototype, "viewChildIndeterminateCb", void 0);
    __decorate([
        core_1.ViewChild('viewChildTestCb'),
        __metadata("design:type", terra_checkbox_component_1.TerraCheckboxComponent)
    ], TerraCheckboxComponentExample.prototype, "viewChildTestCb", void 0);
    TerraCheckboxComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-checkbox-example',
            styles: [require('./terra-checkbox.component.example.scss')],
            template: require('./terra-checkbox.component.example.html'),
        })
    ], TerraCheckboxComponentExample);
    return TerraCheckboxComponentExample;
}());
exports.TerraCheckboxComponentExample = TerraCheckboxComponentExample;
