// @plentymarkets/terra-components/app/components/tree/example/terra-tree.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraTreeComponentExample = /** @class */ (function () {
    function TerraTreeComponentExample() {
        this.leafList = [];
    }
    TerraTreeComponentExample.prototype.ngOnInit = function () {
        this.leafList.push({
            caption: 'Basic-Settings',
            icon: 'icon-settings',
            subLeafList: [
                {
                    caption: 'User',
                    subLeafList: [
                        {
                            caption: 'Accounts'
                        },
                        {
                            caption: 'Role'
                        }
                    ]
                },
                {
                    caption: 'API'
                }
            ]
        });
        this.leafList.push({
            caption: 'Client (Store)',
            subLeafList: [
                {
                    caption: 'Global-Settings'
                },
                {
                    caption: 'Standard'
                }
            ]
        });
        this.leafList.push({
            caption: 'Item',
            icon: 'icon-item',
            subLeafList: [
                {
                    caption: 'Basic-Settings'
                },
                {
                    caption: 'Item availability'
                }
            ]
        });
    };
    TerraTreeComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-tree-example',
            template: require('./terra-tree.component.example.html'),
            styles: [require('./terra-tree.component.example.scss')]
        })
    ], TerraTreeComponentExample);
    return TerraTreeComponentExample;
}());
exports.TerraTreeComponentExample = TerraTreeComponentExample;
