// @plentymarkets/terra-components/app/components/forms/input/color-picker/terra-color-picker.component.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var terra_input_component_1 = require("../terra-input.component");
var color_helper_1 = require("./color.helper");
var terra_regex_1 = require("../../../../helpers/regex/terra-regex");
var TerraColorPickerComponent = /** @class */ (function (_super) {
    __extends(TerraColorPickerComponent, _super);
    function TerraColorPickerComponent() {
        return _super.call(this, terra_regex_1.TerraRegex.COLOR_HEX) || this;
    }
    TerraColorPickerComponent_1 = TerraColorPickerComponent;
    Object.defineProperty(TerraColorPickerComponent.prototype, "color", {
        get: function () {
            return this.value || '#ffffff';
        },
        set: function (c) {
            this.value = c;
        },
        enumerable: true,
        configurable: true
    });
    TerraColorPickerComponent.prototype.isDark = function () {
        return new color_helper_1.Color(this.color).isDark();
    };
    TerraColorPickerComponent = TerraColorPickerComponent_1 = __decorate([
        core_1.Component({
            selector: 'terra-color-picker',
            styles: [require('./terra-color-picker.component.scss')],
            template: require('./terra-color-picker.component.html'),
            providers: [
                {
                    provide: forms_1.NG_VALUE_ACCESSOR,
                    useExisting: core_1.forwardRef(function () { return TerraColorPickerComponent_1; }),
                    multi: true
                }
            ]
        }),
        __metadata("design:paramtypes", [])
    ], TerraColorPickerComponent);
    return TerraColorPickerComponent;
    var TerraColorPickerComponent_1;
}(terra_input_component_1.TerraInputComponent));
exports.TerraColorPickerComponent = TerraColorPickerComponent;
