// @plentymarkets/terra-components/app/components/pager/example/terra-pager.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraPagerComponentExample = /** @class */ (function () {
    function TerraPagerComponentExample() {
    }
    TerraPagerComponentExample.prototype.ngOnInit = function () {
        this.pagingData = {
            pagingUnit: 'pagingEntries',
            totalsCount: 345,
            page: 1,
            itemsPerPage: 50,
            lastPageNumber: 7,
            firstOnPage: 1,
            lastOnPage: 50,
            isLastPage: false
        };
    };
    TerraPagerComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-pager-example',
            styles: [require('./terra-pager.component.example.scss')],
            template: require('./terra-pager.component.example.html'),
        })
    ], TerraPagerComponentExample);
    return TerraPagerComponentExample;
}());
exports.TerraPagerComponentExample = TerraPagerComponentExample;
