// @plentymarkets/terra-components/app/components/layouts/taglist/example/terra-taglist.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraTaglistComponentExample = /** @class */ (function () {
    function TerraTaglistComponentExample() {
        this.infoBoxTagList = [];
        this.infoBoxTagList.push({
            badge: 'Terra'
        }, {
            badge: 'Terra',
            color: 'red'
        }, {
            badge: 'Terra',
            color: '#f3f3f3'
        }, {
            badge: 'Terra',
            isTaggable: true
        }, {
            badge: 'Terra',
            isTaggable: true,
            isTagged: true
        });
    }
    TerraTaglistComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-taglist-example',
            styles: [require('./terra-taglist.component.example.scss')],
            template: require('./terra-taglist.component.example.html'),
        }),
        __metadata("design:paramtypes", [])
    ], TerraTaglistComponentExample);
    return TerraTaglistComponentExample;
}());
exports.TerraTaglistComponentExample = TerraTaglistComponentExample;
