// @plentymarkets/terra-components/app/components/forms/input/example/terra-input.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraInputComponentExample = /** @class */ (function () {
    function TerraInputComponentExample() {
    }
    TerraInputComponentExample.prototype.ngOnInit = function () {
        this.stateSelection = [];
        this.stateSelection.push({
            value: 'hessen',
            caption: 'Hessen'
        }, {
            value: 'bayern',
            caption: 'Bayern'
        }, {
            value: 'sachsen',
            caption: 'Sachsen'
        }, {
            value: 'sachsen-anhalt',
            caption: 'Sachsen Anhalt'
        }, {
            value: 'saarland',
            caption: 'Saarland'
        }, {
            value: 'nordrhein westfalen',
            caption: 'Nordrhein Westfalen'
        }, {
            value: 'rheinland pfalz',
            caption: 'Rheinland Pfalz'
        }, {
            value: 'nordrhein westfalen',
            caption: 'Nordrhein Westfalen'
        }, {
            value: 'niedersachsen',
            caption: 'Niedersachsen'
        }, {
            value: 'baden-w端rttemberg',
            caption: 'Baden-W端rttemberg'
        }, {
            value: 'th端ringen',
            caption: 'Th端ringen'
        }, {
            value: 'berlin',
            caption: 'Berlin'
        }, {
            value: 'brandenburg',
            caption: 'Brandenburg'
        }, {
            value: 'hamburg',
            caption: 'Hamburg'
        }, {
            value: 'bremen',
            caption: 'Bremen'
        }, {
            value: 'schleswig-holstein',
            caption: 'Schleswig-Holstein'
        }, {
            value: 'mecklenburg-vorpommern',
            caption: 'Mecklenburg-Vorpommern'
        });
    };
    TerraInputComponentExample.prototype.showValues = function () {
        alert(this.name + ' ' + this.lastName + ' ' +
            this.email + ' ' + this.password + ' ' +
            this.passwordRepeat + ' ' + this.birthday + ' ' +
            this.state + ' ' + this.address + ' ' +
            this.zip + ' ' + this.city + ' ' +
            this.newsletter + ' ' + this.agbs);
    };
    TerraInputComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-input-example',
            styles: [require('./terra-input.component.example.scss')],
            template: require('./terra-input.component.example.html'),
        })
    ], TerraInputComponentExample);
    return TerraInputComponentExample;
}());
exports.TerraInputComponentExample = TerraInputComponentExample;
