// @plentymarkets/terra-components/app/components/tree/checkbox-tree/example/terra-checkbox-tree.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraCheckboxTreeComponentExample = /** @class */ (function () {
    function TerraCheckboxTreeComponentExample() {
        this.checkboxLeafList = [];
    }
    TerraCheckboxTreeComponentExample.prototype.ngOnInit = function () {
        this.checkboxLeafList.push({
            caption: 'Leaf1',
            icon: 'icon-settings',
            value: 'leaf1Value',
            subLeafList: [
                {
                    caption: 'SubLeaf1',
                    value: 'subLeaf1Value',
                    subLeafList: [
                        {
                            caption: 'SubSubLeaf1',
                            value: 'subSubLeaf1Value',
                        },
                        {
                            caption: 'SubSubLeaf2',
                            value: 'subSubLeaf2Value',
                        }
                    ]
                },
                { caption: 'SubLeaf2' }
            ]
        });
        this.checkboxLeafList.push({
            caption: 'Leaf2',
            icon: 'icon-settings',
            value: 'leaf2Value',
        });
        this.checkboxLeafList.push({
            caption: 'Leaf3',
            value: 'leaf3Value',
        });
    };
    TerraCheckboxTreeComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-checkbox-tree-example',
            template: require('./terra-checkbox-tree.component.example.html'),
            styles: [require('./terra-checkbox-tree.component.example.scss')]
        })
    ], TerraCheckboxTreeComponentExample);
    return TerraCheckboxTreeComponentExample;
}());
exports.TerraCheckboxTreeComponentExample = TerraCheckboxTreeComponentExample;
