// @plentymarkets/terra-components/app/components/editors/syntax-editor/terra-syntax-editor.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var ng2_ace_editor_1 = require("ng2-ace-editor");
require("brace");
require("brace/theme/chrome");
require("brace/mode/typescript");
require("brace/mode/css");
require("brace/mode/javascript");
require("brace/mode/json");
require("brace/mode/scss");
require("brace/mode/html");
require("brace/mode/markdown");
require("brace/mode/twig");
require("brace/mode/php");
require("brace/mode/text");
require("brace/ext/error_marker");
var TerraSyntaxEditorComponent = /** @class */ (function () {
    function TerraSyntaxEditorComponent() {
        this.inputOptions = {
            maxLines: 10000
        };
    }
    TerraSyntaxEditorComponent.prototype.ngAfterViewInit = function () {
        this.editor.getEditor().clearSelection();
        this.editor.getEditor().$blockScrolling = Infinity;
        this.editor.getEditor().setShowPrintMargin(false);
    };
    TerraSyntaxEditorComponent.prototype.setAnnotationList = function (list) {
        this.editor.getEditor().getSession().setAnnotations(list);
    };
    Object.defineProperty(TerraSyntaxEditorComponent.prototype, "inputEditorMode", {
        set: function (value) {
            this._inputEditorMode = value;
            this.editor.setMode(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraSyntaxEditorComponent.prototype, "inputText", {
        get: function () {
            return this._inputText;
        },
        set: function (value) {
            this._inputText = value;
            this.editor.setText(value);
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.ViewChild('aceEditor'),
        __metadata("design:type", ng2_ace_editor_1.AceEditorComponent)
    ], TerraSyntaxEditorComponent.prototype, "editor", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraSyntaxEditorComponent.prototype, "inputReadOnly", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraSyntaxEditorComponent.prototype, "inputOptions", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], TerraSyntaxEditorComponent.prototype, "inputEditorMode", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], TerraSyntaxEditorComponent.prototype, "inputText", null);
    TerraSyntaxEditorComponent = __decorate([
        core_1.Component({
            selector: 'terra-syntax-editor',
            template: require('./terra-syntax-editor.component.html'),
            styles: [require('./terra-syntax-editor.component.scss')]
        }),
        __metadata("design:paramtypes", [])
    ], TerraSyntaxEditorComponent);
    return TerraSyntaxEditorComponent;
}());
exports.TerraSyntaxEditorComponent = TerraSyntaxEditorComponent;
