// @plentymarkets/terra-components/app/components/layouts/info-box/example/terra-info-box.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraInfoboxComponentExample = /** @class */ (function () {
    function TerraInfoboxComponentExample() {
        var _this = this;
        this.tagList = [];
        this.buttonList = [];
        this.tagList.push({
            badge: 'Terra',
            color: 'green'
        }, {
            badge: 'Plenty',
            color: 'red'
        });
        this.buttonList.push({
            caption: 'Terra',
            icon: 'icon-plugin_not_productive',
            clickFunction: function () { return _this.clickFunction('Terra'); }
        }, {
            caption: 'Terra',
            icon: 'icon-plugin_not_productive',
            isTertiary: true,
            clickFunction: function () { return _this.clickFunction('Terra'); }
        });
    }
    TerraInfoboxComponentExample.prototype.clickFunction = function (clickedButton) {
        alert(clickedButton + ' clicked');
    };
    TerraInfoboxComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-info-box-example',
            styles: [require('./terra-info-box.component.example.scss')],
            template: require('./terra-info-box.component.example.html')
        }),
        __metadata("design:paramtypes", [])
    ], TerraInfoboxComponentExample);
    return TerraInfoboxComponentExample;
}());
exports.TerraInfoboxComponentExample = TerraInfoboxComponentExample;
