// @plentymarkets/terra-components/app/components/tables/data-table/context-menu/directive/terra-data-table-context-menu.directive.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_data_table_context_menu_service_1 = require("../service/terra-data-table-context-menu.service");
/**
 * @author mkunze
 */
var TerraDataTableContextMenuDirective = /** @class */ (function () {
    function TerraDataTableContextMenuDirective(contextMenuService) {
        this.contextMenuService = contextMenuService;
    }
    TerraDataTableContextMenuDirective.prototype.rightClicked = function (event) {
        this.contextMenuService.show.next({
            event: event,
            obj: this.inputLinks
        });
        event.preventDefault();
        event.stopPropagation();
    };
    TerraDataTableContextMenuDirective.prototype.ngOnInit = function () {
        this.contextMenuService.init.next(this.inputLinks);
    };
    __decorate([
        core_1.Input('contextMenu'),
        __metadata("design:type", Array)
    ], TerraDataTableContextMenuDirective.prototype, "inputLinks", void 0);
    __decorate([
        core_1.HostListener('contextmenu', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [MouseEvent]),
        __metadata("design:returntype", void 0)
    ], TerraDataTableContextMenuDirective.prototype, "rightClicked", null);
    TerraDataTableContextMenuDirective = __decorate([
        core_1.Directive({
            selector: '[contextMenu]',
        }),
        __metadata("design:paramtypes", [terra_data_table_context_menu_service_1.TerraDataTableContextMenuService])
    ], TerraDataTableContextMenuDirective);
    return TerraDataTableContextMenuDirective;
}());
exports.TerraDataTableContextMenuDirective = TerraDataTableContextMenuDirective;
