// @plentymarkets/terra-components/app/components/forms/input/text-input/example/terra-text-input.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraTextInputComponentExample = /** @class */ (function () {
    function TerraTextInputComponentExample() {
    }
    TerraTextInputComponentExample.prototype.ngOnInit = function () {
        this.readOnlyFieldText = 'this text is not editable';
    };
    TerraTextInputComponentExample.prototype.showUsername = function () {
        alert('Username: ' + this.username);
    };
    TerraTextInputComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-text-input-example',
            styles: [require('./terra-text-input.component.example.scss')],
            template: require('./terra-text-input.component.example.html'),
        })
    ], TerraTextInputComponentExample);
    return TerraTextInputComponentExample;
}());
exports.TerraTextInputComponentExample = TerraTextInputComponentExample;
