// @plentymarkets/terra-components/app/terra-components.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var http_1 = require("@angular/http");
var platform_browser_dynamic_1 = require("@angular/platform-browser-dynamic");
var ngx_bootstrap_1 = require("ngx-bootstrap");
var angular_l10n_1 = require("angular-l10n");
var ngx_quill_1 = require("ngx-quill");
var terra_components_component_1 = require("./terra-components.component");
var platform_browser_1 = require("@angular/platform-browser");
var animations_1 = require("@angular/platform-browser/animations");
var common_1 = require("@angular/common");
var mydatepicker_1 = require("mydatepicker");
var ng2_ace_editor_1 = require("ng2-ace-editor");
var interact_module_1 = require("./components/interactables/interact.module");
var terra_navigator_split_view_config_1 = require("./components/navigator/config/terra-navigator-split-view.config");
var terra_frontend_storage_service_1 = require("./components/file-browser/terra-frontend-storage.service");
var terra_file_browser_service_1 = require("./components/file-browser/terra-file-browser.service");
var l10n_config_1 = require("./translation/l10n.config");
var http_2 = require("@angular/common/http");
var router_1 = require("@angular/router");
var component_collection_1 = require("./components/component-collection");
var example_collection_1 = require("./components/example-collection");
var service_collection_1 = require("./service/service-collection");
var directive_collection_1 = require("./components/directive-collection");
function createCompiler(compilerFactory) {
    return compilerFactory.createCompiler();
}
var TerraComponentsModule = /** @class */ (function () {
    function TerraComponentsModule(l10nLoader) {
        this.l10nLoader = l10nLoader;
        this.l10nLoader.load();
    }
    TerraComponentsModule_1 = TerraComponentsModule;
    TerraComponentsModule.forRoot = function () {
        return {
            ngModule: TerraComponentsModule_1,
            providers: service_collection_1.services
        };
    };
    TerraComponentsModule.forChild = function () {
        return TerraComponentsModule_1;
    };
    TerraComponentsModule = TerraComponentsModule_1 = __decorate([
        core_1.NgModule({
            declarations: [terra_components_component_1.TerraComponentsComponent].concat(component_collection_1.components, directive_collection_1.directives, example_collection_1.examples),
            entryComponents: component_collection_1.components,
            exports: component_collection_1.components.concat(directive_collection_1.directives, example_collection_1.examples),
            imports: [
                platform_browser_1.BrowserModule,
                animations_1.BrowserAnimationsModule,
                common_1.CommonModule,
                forms_1.FormsModule,
                forms_1.ReactiveFormsModule,
                ngx_bootstrap_1.ModalModule.forRoot(),
                http_1.HttpModule,
                http_2.HttpClientModule,
                ngx_bootstrap_1.TooltipModule.forRoot(),
                ngx_bootstrap_1.AlertModule.forRoot(),
                ngx_bootstrap_1.ButtonsModule.forRoot(),
                angular_l10n_1.TranslationModule.forRoot(l10n_config_1.l10nConfig),
                mydatepicker_1.MyDatePickerModule,
                ng2_ace_editor_1.AceEditorModule,
                interact_module_1.TerraInteractModule,
                ngx_quill_1.QuillModule,
                router_1.RouterModule
            ],
            providers: [
                terra_navigator_split_view_config_1.TerraNavigatorSplitViewConfig,
                terra_frontend_storage_service_1.TerraFrontendStorageService,
                terra_file_browser_service_1.TerraFileBrowserService,
                {
                    provide: core_1.COMPILER_OPTIONS,
                    useValue: {},
                    multi: true
                },
                {
                    provide: core_1.CompilerFactory,
                    useClass: platform_browser_dynamic_1.JitCompilerFactory,
                    deps: [core_1.COMPILER_OPTIONS]
                },
                {
                    provide: core_1.Compiler,
                    useFactory: createCompiler,
                    deps: [core_1.CompilerFactory]
                }
            ],
            bootstrap: [
                terra_components_component_1.TerraComponentsComponent
            ]
        }),
        __metadata("design:paramtypes", [angular_l10n_1.L10nLoader])
    ], TerraComponentsModule);
    return TerraComponentsModule;
    var TerraComponentsModule_1;
}());
exports.TerraComponentsModule = TerraComponentsModule;
