// @plentymarkets/terra-components/app/components/buttons/button-with-options/terra-button-with-options.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var util_1 = require("util");
var terra_button_component_1 = require("../button/terra-button.component");
var terra_text_align_enum_1 = require("../../tables/data-table/cell/terra-text-align.enum");
var terra_placement_enum_1 = require("../../../helpers/enums/terra-placement.enum");
var TerraButtonWithOptionsComponent = /** @class */ (function () {
    function TerraButtonWithOptionsComponent(elementRef) {
        var _this = this;
        this.elementRef = elementRef;
        // define click listener
        this.clickListener = function (event) {
            // check if it has been clicked elsewhere
            if (!_this.elementRef.nativeElement.contains(event.target)) {
                _this.optionsToggle = false;
            }
        };
        this.inputOptionsAlignment = terra_text_align_enum_1.TerraTextAlignEnum.RIGHT;
        this.alignRight = true;
    }
    TerraButtonWithOptionsComponent.prototype.ngOnInit = function () {
        // handle undefined input error
        if (util_1.isNullOrUndefined(this.inputOptions)) {
            this.inputOptions = [];
        }
        switch (this.inputOptionsAlignment) {
            case terra_text_align_enum_1.TerraTextAlignEnum.LEFT:
                {
                    this.alignRight = false;
                    break;
                }
            case terra_text_align_enum_1.TerraTextAlignEnum.RIGHT:
                {
                    this.alignRight = true;
                    break;
                }
            case terra_text_align_enum_1.TerraTextAlignEnum.CENTER:
                {
                    this.alignRight = false;
                    break;
                }
        }
    };
    TerraButtonWithOptionsComponent.prototype.optionsClick = function (option) {
        if (!option.isDisabled) {
            // execute click function of the option
            option.clickFunction();
            // close dropdown
            this.toggleOptions();
        }
    };
    TerraButtonWithOptionsComponent.prototype.toggleOptions = function () {
        if (!this.optionsToggle) {
            document.addEventListener('click', this.clickListener);
        }
        else {
            document.removeEventListener('click', this.clickListener);
        }
        this.optionsToggle = !this.optionsToggle;
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraButtonWithOptionsComponent.prototype, "inputCaption", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraButtonWithOptionsComponent.prototype, "inputIcon", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraButtonWithOptionsComponent.prototype, "inputIsPrimary", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraButtonWithOptionsComponent.prototype, "inputIsSecondary", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraButtonWithOptionsComponent.prototype, "inputIsTertiary", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraButtonWithOptionsComponent.prototype, "inputIsDisabled", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraButtonWithOptionsComponent.prototype, "inputTooltipText", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraButtonWithOptionsComponent.prototype, "inputTooltipPlacement", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array)
    ], TerraButtonWithOptionsComponent.prototype, "inputOptions", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraButtonWithOptionsComponent.prototype, "inputOptionsAlignment", void 0);
    __decorate([
        core_1.ViewChild(terra_button_component_1.TerraButtonComponent),
        __metadata("design:type", terra_button_component_1.TerraButtonComponent)
    ], TerraButtonWithOptionsComponent.prototype, "toggleButton", void 0);
    TerraButtonWithOptionsComponent = __decorate([
        core_1.Component({
            selector: 'terra-button-with-options',
            template: require('./terra-button-with-options.component.html'),
            styles: [require('./terra-button-with-options.component.scss')]
        }),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], TerraButtonWithOptionsComponent);
    return TerraButtonWithOptionsComponent;
}());
exports.TerraButtonWithOptionsComponent = TerraButtonWithOptionsComponent;
