// @plentymarkets/terra-components/app/components/note/example/terra-note.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_overlay_component_1 = require("../../layouts/overlay/terra-overlay.component");
var TerraNoteComponentExample = /** @class */ (function () {
    function TerraNoteComponentExample() {
    }
    TerraNoteComponentExample.prototype.ngOnInit = function () {
        var defaultText = 'It is a period of civil war. Rebel spaceships, striking from a hidden base, have won their first ' +
            'victory against the evil Galactic Empire. During the battle, Rebel spies managed to steal secret ' +
            'plans to the Empire\'s ultimate weapon, the Death Star, an armoured space station with enough power ' +
            'to destroy an entire planet. Pursued by the Empire\'s sinister agents, Princess Leia races home aboard ' +
            'her starship, custodian of the stolen plans that can save her people and restore freedom to the galaxy...';
        this.noteTextAndID = defaultText;
        this.noteTextAndSelected = defaultText;
        this.noteTextDynamicExample = defaultText;
        this.editorText = this.noteTextDynamicExample;
    };
    TerraNoteComponentExample.prototype.showOverlay = function () {
        this.overlay.showOverlay();
    };
    TerraNoteComponentExample.prototype.saveText = function (text) {
        this.noteTextDynamicExample = text;
        this.overlay.hideOverlay();
    };
    __decorate([
        core_1.ViewChild('overlay'),
        __metadata("design:type", terra_overlay_component_1.TerraOverlayComponent)
    ], TerraNoteComponentExample.prototype, "overlay", void 0);
    TerraNoteComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-note-example',
            styles: [require('./terra-note.component.example.scss')],
            template: require('./terra-note.component.example.html')
        })
    ], TerraNoteComponentExample);
    return TerraNoteComponentExample;
}());
exports.TerraNoteComponentExample = TerraNoteComponentExample;
