// @plentymarkets/terra-components/app/components/forms/input/number-input/terra-number-input.component.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_input_component_1 = require("../terra-input.component");
var forms_1 = require("@angular/forms");
var terra_regex_1 = require("../../../../helpers/regex/terra-regex");
var nextId = 0;
var TerraNumberInputComponent = /** @class */ (function (_super) {
    __extends(TerraNumberInputComponent, _super);
    function TerraNumberInputComponent() {
        var _this = _super.call(this, terra_regex_1.TerraRegex.NUMERIC) || this;
        // generate the id of the input instance
        _this.id = "number-input_#" + nextId++;
        return _this;
    }
    TerraNumberInputComponent_1 = TerraNumberInputComponent;
    Object.defineProperty(TerraNumberInputComponent.prototype, "inputValue", {
        /**
         * @deprecated inputValue is deprecated. It will be removed in one of the upcoming releases. Please use ngModel instead.
         */
        set: function (v) {
            console.warn('inputValue is deprecated. It will be removed in one of the upcoming releases. Please use ngModel instead.');
            this.value = v;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Set the focus on the native input element.
     */
    TerraNumberInputComponent.prototype.focusNativeInput = function () {
        var _this = this;
        setTimeout(function () {
            var input = document.getElementById(_this.id);
            input.focus();
        });
    };
    /**
     * Select the content of the native input element.
     */
    TerraNumberInputComponent.prototype.selectNativeInput = function () {
        var _this = this;
        setTimeout(function () {
            var input = document.getElementById(_this.id);
            input.select();
        });
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], TerraNumberInputComponent.prototype, "inputValue", null);
    TerraNumberInputComponent = TerraNumberInputComponent_1 = __decorate([
        core_1.Component({
            selector: 'terra-number-input',
            styles: [require('./terra-number-input.component.scss')],
            template: require('./terra-number-input.component.html'),
            providers: [
                {
                    provide: forms_1.NG_VALUE_ACCESSOR,
                    useExisting: core_1.forwardRef(function () { return TerraNumberInputComponent_1; }),
                    multi: true
                }
            ]
        }),
        __metadata("design:paramtypes", [])
    ], TerraNumberInputComponent);
    return TerraNumberInputComponent;
    var TerraNumberInputComponent_1;
}(terra_input_component_1.TerraInputComponent));
exports.TerraNumberInputComponent = TerraNumberInputComponent;
