// @plentymarkets/terra-components/app/components/alert/terra-alert-panel.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_alert_component_1 = require("./terra-alert.component");
/**
 * @author mkunze
 */
var TerraAlertPanelComponent = /** @class */ (function () {
    function TerraAlertPanelComponent() {
        this.alert = terra_alert_component_1.TerraAlertComponent.getInstance();
        this.alerts = this.alert.alerts;
    }
    TerraAlertPanelComponent.prototype.closeAlert = function (i) {
        this.alert.closeAlert(i);
    };
    TerraAlertPanelComponent = __decorate([
        core_1.Component({
            selector: 'terra-alert-panel',
            styles: [
                require('./terra-alert-panel.component.scss'),
                require('./terra-alert-panel.component.glob.scss').toString()
            ],
            template: require('./terra-alert-panel.component.html')
        }),
        __metadata("design:paramtypes", [])
    ], TerraAlertPanelComponent);
    return TerraAlertPanelComponent;
}());
exports.TerraAlertPanelComponent = TerraAlertPanelComponent;
