// @plentymarkets/terra-components/app/components/forms/input/time-picker/terra-time-picker.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @author twieder
 */
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var util_1 = require("util");
var TerraTimePickerComponent = /** @class */ (function () {
    function TerraTimePickerComponent() {
        this.onTouchedCallback = function () { return undefined; };
        this.onChangeCallback = function (_) { return undefined; };
        this.valuesHours = [];
        this.valuesMinutes = [];
        this.value = new Date();
        this.inputIsDisabled = false;
    }
    TerraTimePickerComponent_1 = TerraTimePickerComponent;
    TerraTimePickerComponent.prototype.ngOnInit = function () {
        this.createTimeValues();
    };
    TerraTimePickerComponent.prototype.createTimeValues = function () {
        var hours;
        var minutes;
        for (hours = 0; hours <= 23; hours++) {
            this.valuesHours.push({
                value: hours,
                caption: hours
            });
        }
        for (minutes = 0; minutes <= 59; minutes++) {
            this.valuesMinutes.push({
                value: minutes,
                caption: minutes
            });
        }
    };
    TerraTimePickerComponent.prototype.registerOnChange = function (fn) {
        this.onChangeCallback = fn;
    };
    TerraTimePickerComponent.prototype.registerOnTouched = function (fn) {
        this.onTouchedCallback = fn;
    };
    TerraTimePickerComponent.prototype.writeValue = function (value) {
        this.value = value;
    };
    Object.defineProperty(TerraTimePickerComponent.prototype, "minutes", {
        get: function () {
            if (!util_1.isNullOrUndefined(this.value)) {
                return this.value.getMinutes();
            }
            return 0;
        },
        set: function (minutes) {
            if (!util_1.isNullOrUndefined(this.value)) {
                this.value.setMinutes(minutes);
            }
            this.onChangeCallback(this.value);
            this.onTouchedCallback();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraTimePickerComponent.prototype, "hours", {
        get: function () {
            if (!util_1.isNullOrUndefined(this.value)) {
                return this.value.getHours();
            }
            return 0;
        },
        set: function (minutes) {
            if (!util_1.isNullOrUndefined(this.value)) {
                this.value.setHours(minutes);
            }
            this.onChangeCallback(this.value);
            this.onTouchedCallback();
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraTimePickerComponent.prototype, "inputIsDisabled", void 0);
    TerraTimePickerComponent = TerraTimePickerComponent_1 = __decorate([
        core_1.Component({
            selector: 'terra-time-picker',
            styles: [
                require('./terra-time-picker.component.scss'),
            ],
            template: require('./terra-time-picker.component.html'),
            providers: [
                {
                    provide: forms_1.NG_VALUE_ACCESSOR,
                    useExisting: core_1.forwardRef(function () { return TerraTimePickerComponent_1; }),
                    multi: true
                }
            ]
        }),
        __metadata("design:paramtypes", [])
    ], TerraTimePickerComponent);
    return TerraTimePickerComponent;
    var TerraTimePickerComponent_1;
}());
exports.TerraTimePickerComponent = TerraTimePickerComponent;
