// @plentymarkets/terra-components/app/components/layouts/tag/terra-tag.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var util_1 = require("util");
var color_helper_1 = require("../../forms/input/color-picker/color.helper");
var angular_l10n_1 = require("angular-l10n");
var TerraTagComponent = /** @class */ (function () {
    function TerraTagComponent() {
        this.onCloseTag = new core_1.EventEmitter();
        this.inputIsTagged = false;
        this.inputIsTaggable = false;
        this.inputColor = null;
        this.inputCustomClass = null;
        this.isClosable = false;
    }
    TerraTagComponent.prototype.close = function () {
        this.onCloseTag.emit(this.tagId);
    };
    TerraTagComponent.prototype.getName = function () {
        return this.inputBadge ? this.inputBadge : this.getTranslatedName();
    };
    TerraTagComponent.prototype.getTranslatedName = function () {
        var _this = this;
        // Fallback if names not set
        if (util_1.isNullOrUndefined(this.names)) {
            return this.name;
        }
        else {
            var tagName = this.names.find(function (name) { return name.language === _this.lang; });
            // Fallback if no name for this.lang is set
            if (util_1.isNullOrUndefined(tagName)) {
                return this.name;
            }
            else {
                return tagName.name;
            }
        }
    };
    /**
     * Get the background color.
     * @returns {string}
     * @see inputColor
     */
    TerraTagComponent.prototype.getBgColor = function () {
        if (!util_1.isNullOrUndefined(this.inputColor)) {
            return this.inputColor;
        }
        return null;
    };
    /**
     * Get the foreground color.
     */
    TerraTagComponent.prototype.getColor = function () {
        if (!util_1.isNullOrUndefined(this.inputColor)) {
            return (new color_helper_1.Color(this.inputColor)).isDark() ? '#ffffff' : '#000000';
        }
        return null;
    };
    __decorate([
        angular_l10n_1.Language(),
        __metadata("design:type", String)
    ], TerraTagComponent.prototype, "lang", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraTagComponent.prototype, "name", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraTagComponent.prototype, "inputBadge", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraTagComponent.prototype, "inputIsTagged", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraTagComponent.prototype, "inputIsTaggable", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraTagComponent.prototype, "inputCustomClass", void 0);
    __decorate([
        core_1.Input()
        /**
         * The background color for the tag.
         */
        ,
        __metadata("design:type", String)
    ], TerraTagComponent.prototype, "inputColor", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number)
    ], TerraTagComponent.prototype, "tagId", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraTagComponent.prototype, "isClosable", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array)
    ], TerraTagComponent.prototype, "names", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraTagComponent.prototype, "onCloseTag", void 0);
    TerraTagComponent = __decorate([
        core_1.Component({
            selector: 'terra-tag',
            styles: [require('./terra-tag.component.scss')],
            template: require('./terra-tag.component.html')
        }),
        __metadata("design:paramtypes", [])
    ], TerraTagComponent);
    return TerraTagComponent;
}());
exports.TerraTagComponent = TerraTagComponent;
