// @plentymarkets/terra-components/app/components/file-browser/terra-file-browser.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var util_1 = require("util");
var TerraFileBrowserService = /** @class */ (function () {
    function TerraFileBrowserService(zone) {
        this.zone = zone;
        this.isDragActive = new BehaviorSubject_1.BehaviorSubject(false);
        this.dropzones = [];
        this.globalListenersDefined = false;
        this.dragenterTarget = null;
    }
    TerraFileBrowserService.prototype.addDropzone = function (dropzone) {
        if (!this.globalListenersDefined) {
            this.setupGlobalListeners();
        }
        this.dropzones.push(dropzone);
    };
    TerraFileBrowserService.prototype.removeDropzone = function (dropzone) {
        var idx = this.dropzones.indexOf(dropzone);
        if (idx >= 0) {
            this.dropzones.splice(idx, 1);
        }
    };
    TerraFileBrowserService.prototype.isDropzone = function (element) {
        return this.dropzones.some(function (dropzone) {
            return dropzone.contains(element);
        });
    };
    TerraFileBrowserService.prototype.setupGlobalListeners = function () {
        var _this = this;
        this.zone.runOutsideAngular(function () {
            var setEffect = function (event) {
                if (_this.isDropzone(event.target)) {
                    event.dataTransfer.effectAllowed = 'copy';
                    event.dataTransfer.dropEffect = 'copy';
                }
                else {
                    event.dataTransfer.effectAllowed = 'none';
                    event.dataTransfer.dropEffect = 'none';
                }
            };
            var isFileEvent = function (event) {
                return !util_1.isNullOrUndefined(event.dataTransfer.types) && event.dataTransfer.types.indexOf('Files') >= 0;
            };
            window.addEventListener('dragenter', function (event) {
                if (isFileEvent(event)) {
                    _this.dragenterTarget = event.target;
                    event.preventDefault();
                    if (!_this.isDragActive.value) {
                        _this.isDragActive.next(true);
                    }
                }
            });
            window.addEventListener('dragover', function (event) {
                if (isFileEvent(event)) {
                    event.preventDefault();
                    setEffect(event);
                }
            });
            window.addEventListener('dragleave', function (event) {
                if (isFileEvent(event) && event.target === _this.dragenterTarget) {
                    _this.dragenterTarget = null;
                    _this.isDragActive.next(false);
                }
            });
            window.addEventListener('drop', function (event) {
                if (isFileEvent(event)) {
                    event.preventDefault();
                    _this.dragenterTarget = null;
                    if (_this.isDragActive.value) {
                        _this.isDragActive.next(false);
                    }
                }
            });
        });
        this.globalListenersDefined = true;
    };
    TerraFileBrowserService = __decorate([
        core_1.Injectable(),
        __metadata("design:paramtypes", [core_1.NgZone])
    ], TerraFileBrowserService);
    return TerraFileBrowserService;
}());
exports.TerraFileBrowserService = TerraFileBrowserService;
