// @plentymarkets/terra-components/app/components/forms/tag-select/example/terra-tag-select.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
/**
 * @author mfrank
 */
var TerraTagSelectComponentExample = /** @class */ (function () {
    function TerraTagSelectComponentExample() {
        this.selectedTags = [{
                id: 3,
                name: 'Fast schwarz',
                color: '#000350',
                isClosable: true,
                names: [
                    {
                        id: 3,
                        tagId: 3,
                        language: 'de',
                        name: 'Fast schwarz'
                    },
                    {
                        id: 7,
                        tagId: 3,
                        language: 'en',
                        name: 'Almost black'
                    }
                ]
            }];
        this.tags = [
            {
                id: 4,
                name: 'Dunkel grau',
                color: '#989898',
                names: [
                    {
                        id: 4,
                        tagId: 4,
                        language: 'de',
                        name: 'Dunkel grau'
                    },
                    {
                        id: 5,
                        tagId: 4,
                        language: 'en',
                        name: 'Dark grey'
                    }
                ],
            },
            {
                id: 3,
                name: 'Fast schwarz',
                color: '#000350',
                names: [
                    {
                        id: 3,
                        tagId: 3,
                        language: 'de',
                        name: 'Fast schwarz'
                    },
                    {
                        id: 7,
                        tagId: 3,
                        language: 'en',
                        name: 'Almost black'
                    }
                ],
            },
            {
                id: 2,
                name: 'Helles grau',
                color: '#f0eef1',
                names: [
                    {
                        id: 2,
                        tagId: 2,
                        language: 'de',
                        name: 'Helles grau'
                    },
                    {
                        id: 8,
                        tagId: 2,
                        language: 'en',
                        name: 'Light grey'
                    }
                ],
            },
            {
                id: 1,
                name: 'Rot',
                color: '#ff4b31',
                names: [
                    {
                        id: 1,
                        tagId: 1,
                        language: 'de',
                        name: 'Rot'
                    },
                    {
                        id: 6,
                        tagId: 1,
                        language: 'en',
                        name: 'Red'
                    }
                ],
            }
        ];
    }
    TerraTagSelectComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-tag-select-example',
            template: require('./terra-tag-select.component.example.html')
        })
    ], TerraTagSelectComponentExample);
    return TerraTagSelectComponentExample;
}());
exports.TerraTagSelectComponentExample = TerraTagSelectComponentExample;
