// @plentymarkets/terra-components/app/components/layouts/overlay/terra-overlay.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var ngx_bootstrap_1 = require("ngx-bootstrap");
/**
 * @author mfrank
 */
var TerraOverlayComponent = /** @class */ (function () {
    function TerraOverlayComponent() {
        this.inputIsStatic = false;
        this.inputIsCloseable = true;
        this.inputIsExtraLarge = false;
        this.inputIsLarge = false;
        this.inputIsSmall = false;
        this.outputOnHide = new core_1.EventEmitter();
        this.outputOnShow = new core_1.EventEmitter();
    }
    TerraOverlayComponent.prototype.ngAfterViewInit = function () {
        if (!this.inputIsCloseable) {
            this.inputIsStatic = true;
        }
        if (this.inputIsStatic) {
            this.viewChildOverlay.config.backdrop = 'static';
            this.viewChildOverlay.config.keyboard = false;
        }
    };
    TerraOverlayComponent.prototype.showOverlay = function () {
        this.viewChildOverlay.show();
    };
    TerraOverlayComponent.prototype.hideOverlay = function () {
        this.viewChildOverlay.hide();
    };
    TerraOverlayComponent.prototype.emitOutputOnShow = function () {
        this.outputOnShow.emit(null);
    };
    TerraOverlayComponent.prototype.emitOutputOnHide = function () {
        this.outputOnHide.emit(null);
    };
    __decorate([
        core_1.ViewChild('viewChildOverlay'),
        __metadata("design:type", ngx_bootstrap_1.ModalDirective)
    ], TerraOverlayComponent.prototype, "viewChildOverlay", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraOverlayComponent.prototype, "inputOverlayTitle", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraOverlayComponent.prototype, "inputPrimaryButtonInterface", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraOverlayComponent.prototype, "inputSecondaryButtonInterface", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraOverlayComponent.prototype, "inputIsStatic", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraOverlayComponent.prototype, "inputIsCloseable", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraOverlayComponent.prototype, "inputIsLarge", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraOverlayComponent.prototype, "inputIsSmall", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraOverlayComponent.prototype, "inputIsExtraLarge", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraOverlayComponent.prototype, "outputOnHide", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraOverlayComponent.prototype, "outputOnShow", void 0);
    TerraOverlayComponent = __decorate([
        core_1.Component({
            selector: 'terra-overlay',
            template: require('./terra-overlay.component.html'),
            styles: [require('./terra-overlay.component.scss')]
        }),
        __metadata("design:paramtypes", [])
    ], TerraOverlayComponent);
    return TerraOverlayComponent;
}());
exports.TerraOverlayComponent = TerraOverlayComponent;
