// @plentymarkets/terra-components/app/components/nested-data-picker/example/terra-nested-data-picker.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_nested_data_picker_service_example_1 = require("../example/service/terra-nested-data-picker.service.example");
var TerraNestedDataPickerComponentExample = /** @class */ (function () {
    function TerraNestedDataPickerComponentExample(terraService) {
        this.terraService = terraService;
    }
    TerraNestedDataPickerComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-nested-data-picker-example',
            styles: [require('./terra-nested-data-picker.component.example.scss')],
            template: require('./terra-nested-data-picker.component.example.html'),
            providers: [terra_nested_data_picker_service_example_1.NestedPickerExampleService]
        }),
        __metadata("design:paramtypes", [terra_nested_data_picker_service_example_1.NestedPickerExampleService])
    ], TerraNestedDataPickerComponentExample);
    return TerraNestedDataPickerComponentExample;
}());
exports.TerraNestedDataPickerComponentExample = TerraNestedDataPickerComponentExample;
