// @plentymarkets/terra-components/app/components/dynamic-module-loader/terra-dynamic-module-loader.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var util_1 = require("util");
var TerraDynamicModuleLoaderComponent = /** @class */ (function () {
    function TerraDynamicModuleLoaderComponent(jitCompiler) {
        this.jitCompiler = jitCompiler;
    }
    TerraDynamicModuleLoaderComponent.prototype.ngAfterViewInit = function () {
        this.resolvedData = this.inputModule;
        this.updateComponent();
    };
    TerraDynamicModuleLoaderComponent.prototype.ngOnChanges = function (changes) {
        if (changes.hasOwnProperty('inputInputs')) {
            this.assignInputProperties();
        }
    };
    TerraDynamicModuleLoaderComponent.prototype.ngOnDestroy = function () {
        if (this.cmpRef) {
            this.cmpRef.destroy();
        }
    };
    TerraDynamicModuleLoaderComponent.prototype.updateComponent = function () {
        var _this = this;
        this.jitCompiler.compileModuleAndAllComponentsAsync(this.resolvedData.ngModule)
            .then(function (moduleWithFactories) {
            moduleWithFactories.componentFactories.forEach(function (factory) {
                if (_this.inputMainComponentName === factory.componentType.name) {
                    // create the component
                    _this.cmpRef = _this.viewChildTarget.createComponent(factory);
                    // pass the instance of the loaded view back to the component
                    _this.cmpRef.instance.splitViewInstance = _this.inputView;
                    // pass the delivered parameter to the component
                    _this.cmpRef.instance.parameter = _this.inputParameter; // TODO: deprecated if old split view is removed
                    // add inputs to component for data binding purposes
                    _this.assignInputProperties();
                }
            });
        });
    };
    TerraDynamicModuleLoaderComponent.prototype.assignInputProperties = function () {
        var _this = this;
        if (!util_1.isNullOrUndefined(this.inputInputs) && this.cmpRef) {
            this.inputInputs.forEach(function (input) {
                if (!util_1.isNullOrUndefined(input) && !util_1.isNullOrUndefined(input.name)) {
                    _this.cmpRef.instance[input.name] = input.value;
                }
            });
            this.cmpRef.changeDetectorRef.detectChanges();
        }
    };
    __decorate([
        core_1.ViewChild('viewChildTarget', { read: core_1.ViewContainerRef }),
        __metadata("design:type", core_1.ViewContainerRef)
    ], TerraDynamicModuleLoaderComponent.prototype, "viewChildTarget", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraDynamicModuleLoaderComponent.prototype, "inputModule", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraDynamicModuleLoaderComponent.prototype, "inputMainComponentName", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraDynamicModuleLoaderComponent.prototype, "inputParameter", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array)
    ], TerraDynamicModuleLoaderComponent.prototype, "inputInputs", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraDynamicModuleLoaderComponent.prototype, "inputView", void 0);
    TerraDynamicModuleLoaderComponent = __decorate([
        core_1.Component({
            selector: 'terra-dynamic-module-loader',
            template: require('./terra-dynamic-module-loader.component.html'),
            styles: [require('./terra-dynamic-module-loader.component.scss')]
        }),
        __metadata("design:paramtypes", [core_1.Compiler])
    ], TerraDynamicModuleLoaderComponent);
    return TerraDynamicModuleLoaderComponent;
}());
exports.TerraDynamicModuleLoaderComponent = TerraDynamicModuleLoaderComponent;
