// dgeni-packages/typescript/services/TsParser/FileInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Location_1 = require("./Location");
var fs_1 = require("fs");
var path = require('canonical-path');
/**
 * The file (and the location in the file) from where an API doc element was sourced.
 */
var FileInfo = /** @class */ (function () {
    function FileInfo(declaration, basePath) {
        this.declaration = declaration;
        this.basePath = basePath;
        this.relativePath = this.declaration.getSourceFile().fileName;
        this.location = new Location_1.Location(this.declaration);
        this.filePath = path.resolve(this.basePath, this.relativePath);
        this.baseName = path.basename(this.filePath, path.extname(this.filePath));
        this.extension = path.extname(this.filePath).replace(/^\./, '');
        this.projectRelativePath = path.relative(this.basePath, this.filePath);
        this.realFilePath = this.getRealFilePath(this.filePath);
        this.realProjectRelativePath = path.relative(this.basePath, this.realFilePath);
    }
    FileInfo.prototype.getRealFilePath = function (filePath) {
        return fs_1.realpathSync(filePath).replace(RegExp('\\' + path.sep, 'g'), '/');
    };
    return FileInfo;
}());
exports.FileInfo = FileInfo;
