// @plentymarkets/terra-components/app/components/filter/example/terra-filter.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraFilterComponentExample = /** @class */ (function () {
    function TerraFilterComponentExample() {
        this.name = '';
        this.listBoxValues = [];
        this.selectedListBoxValue = 1;
    }
    TerraFilterComponentExample.prototype.ngOnInit = function () {
        for (var i = 1; i < 4; i++) {
            this.listBoxValues.push({
                value: 'test' + i,
                caption: 'Test' + i
            });
        }
    };
    TerraFilterComponentExample.prototype.onSearchBtnClicked = function () {
        alert('filtered for ' + this.selectedListBoxValue);
    };
    TerraFilterComponentExample.prototype.onResetBtnClicked = function () {
        return;
    };
    TerraFilterComponentExample.prototype.onSubmit = function () {
        this.onSearchBtnClicked();
    };
    TerraFilterComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-filter-example',
            styles: [require('./terra-filter.component.example.scss')],
            template: require('./terra-filter.component.example.html'),
        })
    ], TerraFilterComponentExample);
    return TerraFilterComponentExample;
}());
exports.TerraFilterComponentExample = TerraFilterComponentExample;
