// @plentymarkets/terra-components/app/components/layouts/column-container/terra-two-columns-container.directive.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var router_1 = require("@angular/router");
var two_column_helper_1 = require("../../../helpers/two-column.helper");
var terra_two_columns_container_component_1 = require("./terra-two-columns-container.component");
var util_1 = require("util");
var TerraTwoColumnsContainerDirective = /** @class */ (function () {
    function TerraTwoColumnsContainerDirective(route, router, twoColComponent) {
        this.route = route;
        this.router = router;
        this.twoColComponent = twoColComponent;
        this.basePath = router.url;
    }
    TerraTwoColumnsContainerDirective.prototype.ngOnInit = function () {
        var _this = this;
        var subscribable = this.router.events.filter(function (event) {
            return event instanceof router_1.NavigationEnd && event.urlAfterRedirects === _this.basePath;
        });
        this.subscription = subscribable.subscribe(function (event) {
            if (event.url !== event.urlAfterRedirects) {
                _this.setColumnHidden('right');
            }
            else {
                _this.setColumnHidden('left');
            }
        });
        this.setColumnHidden('left');
        this.route.data.subscribe(function (data) {
            _this.basePath = _this.router.url;
        });
    };
    TerraTwoColumnsContainerDirective.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    TerraTwoColumnsContainerDirective.prototype.setColumnHidden = function (column) {
        if (!util_1.isNullOrUndefined(this.twoColComponent)) {
            if (column === 'right') {
                this.twoColComponent.leftColumn = two_column_helper_1.TwoColumnHelper.leftRightColXS()
                    + two_column_helper_1.TwoColumnHelper.leftColMD(this.twoColComponent.leftColumnWidth)
                    + two_column_helper_1.TwoColumnHelper.leftColLG(this.twoColComponent.leftColumnWidth);
                this.twoColComponent.rightColumn = two_column_helper_1.TwoColumnHelper.leftRightHiddenXS()
                    + two_column_helper_1.TwoColumnHelper.rightColMD(this.twoColComponent.leftColumnWidth)
                    + two_column_helper_1.TwoColumnHelper.rightColLG(this.twoColComponent.leftColumnWidth);
            }
            else if (column === 'left') {
                this.twoColComponent.leftColumn = two_column_helper_1.TwoColumnHelper.leftRightHiddenXS()
                    + two_column_helper_1.TwoColumnHelper.leftColMD(this.twoColComponent.leftColumnWidth)
                    + two_column_helper_1.TwoColumnHelper.leftColLG(this.twoColComponent.leftColumnWidth);
                this.twoColComponent.rightColumn = two_column_helper_1.TwoColumnHelper.leftRightColXS()
                    + two_column_helper_1.TwoColumnHelper.rightColMD(this.twoColComponent.leftColumnWidth)
                    + two_column_helper_1.TwoColumnHelper.rightColLG(this.twoColComponent.leftColumnWidth);
            }
        }
    };
    TerraTwoColumnsContainerDirective = __decorate([
        core_1.Directive({
            selector: 'terra-2-col[mobileRouting]'
        }),
        __param(2, core_1.Host()), __param(2, core_1.Optional()),
        __metadata("design:paramtypes", [router_1.ActivatedRoute,
            router_1.Router,
            terra_two_columns_container_component_1.TerraTwoColumnsContainerComponent])
    ], TerraTwoColumnsContainerDirective);
    return TerraTwoColumnsContainerDirective;
}());
exports.TerraTwoColumnsContainerDirective = TerraTwoColumnsContainerDirective;
