// @plentymarkets/terra-components/app/components/nested-data-picker/example/service/terra-nested-data-picker.service.example.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var Observable_1 = require("rxjs/Observable");
var util_1 = require("util");
var terra_nested_data_picker_base_service_1 = require("../../../nested-data-picker/service/terra-nested-data-picker-base.service");
var NestedPickerExampleService = /** @class */ (function (_super) {
    __extends(NestedPickerExampleService, _super);
    function NestedPickerExampleService() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.parents = {
            entries: [
                {
                    id: 1,
                    hasChildren: true,
                    details: [
                        {
                            detailId: 1,
                            name: 'Parent 1'
                        }
                    ]
                },
                {
                    id: 2,
                    hasChildren: true,
                    details: [
                        {
                            detailId: 2,
                            name: 'Parent 2'
                        }
                    ]
                },
                {
                    id: 3,
                    hasChildren: true,
                    details: [
                        {
                            detailId: 3,
                            name: 'Parent 3'
                        }
                    ]
                }
            ],
            firstOnPage: 1,
            isLastPage: true,
            itemsPerPage: 50,
            lastOnPage: 3,
            lastPageNumber: 1,
            page: 1,
            totalsCount: 3
        };
        _this.childs = {
            entries: [
                {
                    id: 1,
                    parentId: 1,
                    details: [
                        {
                            detailId: 1,
                            name: 'Child 1'
                        }
                    ]
                },
                {
                    id: 2,
                    parentId: 1,
                    details: [
                        {
                            detailId: 1,
                            name: 'Child 2'
                        }
                    ]
                },
                {
                    id: 3,
                    parentId: 2,
                    details: [
                        {
                            detailId: 2,
                            name: 'Child 3'
                        }
                    ]
                },
                {
                    id: 4,
                    parentId: 3,
                    details: [
                        {
                            detailId: 3,
                            name: 'Child 4'
                        }
                    ]
                }
            ],
            firstOnPage: 1,
            isLastPage: true,
            itemsPerPage: 50,
            lastOnPage: 3,
            lastPageNumber: 1,
            page: 1,
            totalsCount: 3
        };
        return _this;
    }
    NestedPickerExampleService.prototype.requestNestedData = function (parentId) {
        var data;
        if (!util_1.isNullOrUndefined(parentId)) {
            data = {
                entries: [],
                firstOnPage: 1,
                isLastPage: true,
                itemsPerPage: 50,
                lastOnPage: 3,
                lastPageNumber: 1,
                page: 1,
                totalsCount: 3
            };
            this.childs.entries.forEach(function (child) {
                if (child.parentId === parentId) {
                    data.entries.push(child);
                }
            });
        }
        else {
            data = this.parents;
        }
        return Observable_1.Observable.create(function (observer) {
            observer.next(data);
            observer.complete();
        });
    };
    NestedPickerExampleService.prototype.requestNestedDataById = function (id) {
        var childs;
        childs = {
            entries: [],
            firstOnPage: 1,
            isLastPage: true,
            itemsPerPage: 50,
            lastOnPage: 3,
            lastPageNumber: 1,
            page: 1,
            totalsCount: 3
        };
        this.childs.entries.forEach(function (child) {
            if (child.parentId === id) {
                childs.entries.push(child);
            }
        });
        return Observable_1.Observable.create(function (observer) {
            observer.next(childs);
            observer.complete();
        });
    };
    NestedPickerExampleService = __decorate([
        core_1.Injectable()
    ], NestedPickerExampleService);
    return NestedPickerExampleService;
}(terra_nested_data_picker_base_service_1.TerraNestedDataPickerBaseService));
exports.NestedPickerExampleService = NestedPickerExampleService;
