// @plentymarkets/terra-components/app/components/pager/terra-pager.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var Subject_1 = require("rxjs/Subject");
var TerraPagerComponent = /** @class */ (function () {
    function TerraPagerComponent(zone) {
        this.zone = zone;
        this.outputDoPaging = new core_1.EventEmitter();
        this.pagingClicks = new Subject_1.Subject();
    }
    TerraPagerComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.pagingClicks.debounceTime(500).subscribe(function (e) { return _this.outputDoPaging.emit(e); });
        if (!this.inputDefaultPagingSize) {
            this.inputDefaultPagingSize = 25;
        }
        if (!this.inputPagingSize) {
            this.inputPagingSize = [
                {
                    value: 25,
                    caption: '25'
                },
                {
                    value: 50,
                    caption: '50'
                },
                {
                    value: 75,
                    caption: '75'
                },
                {
                    value: 100,
                    caption: '100'
                }
            ];
        }
        if (!this.inputPagingData) {
            this.inputPagingData = {
                pagingUnit: '',
                totalsCount: 0,
                page: 1,
                itemsPerPage: 0,
                lastPageNumber: 0,
                firstOnPage: 0,
                lastOnPage: 0,
                isLastPage: false
            };
        }
    };
    TerraPagerComponent.prototype.onFirstPage = function () {
        if (!this.inputRequestPending) {
            this.inputPagingData.page = 1;
        }
        this.notify();
    };
    TerraPagerComponent.prototype.onPrevPage = function () {
        if (!this.inputRequestPending) {
            this.inputPagingData.page -= 1;
        }
        this.notify();
    };
    TerraPagerComponent.prototype.onNextPage = function () {
        if (!this.inputRequestPending) {
            this.inputPagingData.page += 1;
        }
        this.notify();
    };
    TerraPagerComponent.prototype.onLastPage = function () {
        if (!this.inputRequestPending) {
            this.inputPagingData.page = this.inputPagingData.lastPageNumber;
        }
        this.notify();
    };
    TerraPagerComponent.prototype.onReload = function () {
        this.notify();
    };
    TerraPagerComponent.prototype.onToPage = function (event, pageNumber) {
        event.preventDefault();
        // Limit page number to valid range [1, lastPageNumber]
        this.inputPagingData.page = Math.max(1, Math.min(this.inputPagingData.lastPageNumber, pageNumber));
        this.outputDoPaging.emit(this.inputPagingData);
    };
    TerraPagerComponent.prototype.onChangeOffsetTo = function (selectedOffset) {
        this.inputPagingData.page = 1;
        this.inputPagingData.itemsPerPage = selectedOffset.value;
        this.outputDoPaging.emit(this.inputPagingData);
    };
    TerraPagerComponent.prototype.notify = function () {
        if (!this.inputRequestPending) {
            this.pagingClicks.next(this.inputPagingData);
        }
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraPagerComponent.prototype, "inputPagingData", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number)
    ], TerraPagerComponent.prototype, "inputDefaultPagingSize", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array)
    ], TerraPagerComponent.prototype, "inputPagingSize", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraPagerComponent.prototype, "inputRequestPending", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraPagerComponent.prototype, "outputDoPaging", void 0);
    TerraPagerComponent = __decorate([
        core_1.Component({
            selector: 'terra-pager',
            styles: [require('./terra-pager.component.scss')],
            template: require('./terra-pager.component.html')
        }),
        __metadata("design:paramtypes", [core_1.NgZone])
    ], TerraPagerComponent);
    return TerraPagerComponent;
}());
exports.TerraPagerComponent = TerraPagerComponent;
