// @plentymarkets/terra-components/app/components/tables/simple/example/terra-table-simple.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_simple_table_component_1 = require("../terra-simple-table.component");
var TerraSimpleTableComponentExample = /** @class */ (function () {
    function TerraSimpleTableComponentExample(viewContainerRef) {
        this._headerList = [];
        this._rowList = [];
        // You need this small hack in order to catch application root view container ref
        this.viewContainerRef = viewContainerRef;
    }
    TerraSimpleTableComponentExample.prototype.ngOnInit = function () {
        for (var x = 0; x < 5; x++) {
            var cell = {
                caption: 'header ' + x,
                width: '100',
            };
            this.headerList.push(cell);
        }
        this.headerList.push({
            caption: 'buttons',
            width: '100'
        });
        for (var i = 1; i < 10; i++) {
            var cellList = [];
            for (var j = 0; j < 5; j++) {
                var cell = {
                    caption: 'row' + i + 'testcell ' + j,
                    icon: 'icon-referrer_backend'
                };
                cellList.push(cell);
            }
            var buttonList = [];
            buttonList.push({
                caption: 'hallo',
                clickFunction: function () {
                    alert('test');
                }
            });
            var buttonCell = {
                buttonList: buttonList
            };
            cellList.push(buttonCell);
            var row = {
                cellList: cellList,
                disabled: i % 3 === 0,
                selected: i % 2 === 0
            };
            this.rowList.push(row);
        }
    };
    Object.defineProperty(TerraSimpleTableComponentExample.prototype, "headerList", {
        get: function () {
            return this._headerList;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraSimpleTableComponentExample.prototype, "rowList", {
        get: function () {
            return this._rowList;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.ViewChild('table'),
        __metadata("design:type", terra_simple_table_component_1.TerraSimpleTableComponent)
    ], TerraSimpleTableComponentExample.prototype, "table", void 0);
    TerraSimpleTableComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-simple-table-example',
            template: require('./terra-simple-table.component.example.html'),
            styles: [require('./terra-simple-table.component.example.scss')],
        }),
        __metadata("design:paramtypes", [core_1.ViewContainerRef])
    ], TerraSimpleTableComponentExample);
    return TerraSimpleTableComponentExample;
}());
exports.TerraSimpleTableComponentExample = TerraSimpleTableComponentExample;
