// @plentymarkets/terra-components/app/components/forms/radio-button/terra-radio-button.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
/**
 * @author mfrank
 */
var TerraRadioButtonComponent = /** @class */ (function () {
    function TerraRadioButtonComponent() {
        this.onTouchedCallback = function () { return undefined; };
        this.onChangeCallback = function (_) { return undefined; };
        this.inputIsUncheckable = false;
        this.inputIsDisabled = false;
    }
    TerraRadioButtonComponent_1 = TerraRadioButtonComponent;
    TerraRadioButtonComponent.prototype.onClick = function () {
        if (this.inputIsDisabled) {
            return;
        }
        if (this.inputIsUncheckable && this.inputValue === this.value) {
            this.value = undefined;
        }
        else {
            this.value = this.inputValue;
        }
        this.onTouchedCallback();
        this.onChangeCallback(this.value);
    };
    TerraRadioButtonComponent.prototype.writeValue = function (value) {
        this.value = value;
    };
    TerraRadioButtonComponent.prototype.registerOnChange = function (fn) {
        this.onChangeCallback = fn;
    };
    TerraRadioButtonComponent.prototype.registerOnTouched = function (fn) {
        this.onTouchedCallback = fn;
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraRadioButtonComponent.prototype, "inputCaption", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraRadioButtonComponent.prototype, "inputValue", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraRadioButtonComponent.prototype, "inputIsDisabled", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraRadioButtonComponent.prototype, "inputIsUncheckable", void 0);
    __decorate([
        core_1.HostListener('click'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], TerraRadioButtonComponent.prototype, "onClick", null);
    TerraRadioButtonComponent = TerraRadioButtonComponent_1 = __decorate([
        core_1.Component({
            selector: 'terra-radio-button',
            template: require('./terra-radio-button.component.html'),
            styles: [require('./terra-radio-button.component.scss')],
            providers: [
                {
                    provide: forms_1.NG_VALUE_ACCESSOR,
                    useExisting: core_1.forwardRef(function () { return TerraRadioButtonComponent_1; }),
                    multi: true
                }
            ]
        }),
        __metadata("design:paramtypes", [])
    ], TerraRadioButtonComponent);
    return TerraRadioButtonComponent;
    var TerraRadioButtonComponent_1;
}());
exports.TerraRadioButtonComponent = TerraRadioButtonComponent;
