// @plentymarkets/terra-components/app/components/category-picker/terra-category-picker.component.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var angular_l10n_1 = require("angular-l10n");
var forms_1 = require("@angular/forms");
var terra_category_picker_base_service_1 = require("./service/terra-category-picker-base.service");
var util_1 = require("util");
var terra_nested_data_picker_component_1 = require("../nested-data-picker/terra-nested-data-picker.component");
var nested_data_tree_config_1 = require("../nested-data-picker/config/nested-data-tree.config");
var __1 = require("../../..");
var TerraCategoryPickerComponent = /** @class */ (function (_super) {
    __extends(TerraCategoryPickerComponent, _super);
    function TerraCategoryPickerComponent(translation, nestedTreeConfig) {
        var _this = _super.call(this, translation, nestedTreeConfig) || this;
        _this.translation = translation;
        _this.nestedTreeConfig = nestedTreeConfig;
        _this.value = 0;
        _this.completeCategory = {
            id: null,
            isActive: null,
            isOpen: null,
            isVisible: null,
            name: '',
            tooltip: '',
            tooltipPlacement: '',
        };
        _this.categoryName = '';
        _this.list = [];
        _this.isContainerCategorySelected = false;
        _this.isNotInitialCall = false;
        return _this;
    }
    TerraCategoryPickerComponent_1 = TerraCategoryPickerComponent;
    TerraCategoryPickerComponent.prototype.ngAfterContentChecked = function () {
        if (this.nestedTreeConfig.list.length === 0) {
            this.nestedTreeConfig.list = this.list;
        }
        if (!util_1.isNullOrUndefined(this.nestedTreeConfig.currentSelectedNode) && !util_1.isNullOrUndefined(this.nestedTreeConfig.currentSelectedNode.value)) {
            this.isContainerCategorySelected = (this.nestedTreeConfig.currentSelectedNode.value.data.type === 'container');
        }
    };
    TerraCategoryPickerComponent.prototype.ngOnInit = function () {
        if (util_1.isNullOrUndefined(this.inputName)) {
            this.inputName = this.translation.translate('terraCategoryPicker.category');
        }
        this.nestedTreeConfig.list = this.list;
        this.getCategoriesByParent(null);
    };
    TerraCategoryPickerComponent.prototype.getCompleteCategoryObject = function () {
        return this.completeCategory;
    };
    // From ControlValueAccessor interface
    TerraCategoryPickerComponent.prototype.writeValue = function (value) {
        var _this = this;
        if (!util_1.isNullOrUndefined(value)) {
            this.inputCategoryService.requestCategoryDataById(value).subscribe(function (data) {
                if (util_1.isNullOrUndefined(_this.nestedTreeConfig.findNodeById(value))) {
                    _this.addNodes(data, null);
                }
                var nodeToSelect = _this.nestedTreeConfig.findNodeById(value);
                if (!util_1.isNullOrUndefined(nodeToSelect)) {
                    _this.nestedTreeConfig.currentSelectedNode = nodeToSelect;
                    _this.categoryName = _this.nestedTreeConfig.currentSelectedNode.name;
                }
                _this.value = value;
                if (_this.isNotInitialCall) {
                    _this.updateCompleteCategory(nodeToSelect);
                    _this.onTouchedCallback();
                    _this.onChangeCallback(_this.value);
                }
            });
        }
    };
    TerraCategoryPickerComponent.prototype.onSelectNode = function () {
        this.isNotInitialCall = true;
        if (!util_1.isNullOrUndefined(this.nestedTreeConfig.currentSelectedNode)) {
            this.categoryName = this.nestedTreeConfig.currentSelectedNode.name;
            this.writeValue(this.nestedTreeConfig.currentSelectedNode.id);
        }
        this.toggleTree = !this.toggleTree;
    };
    TerraCategoryPickerComponent.prototype.reset = function () {
        this.nestedTreeConfig.currentSelectedNode = {
            id: null,
            isActive: null,
            isOpen: null,
            isVisible: null,
            name: '',
            tooltip: '',
            tooltipPlacement: '',
        };
        this.categoryName = '';
        this.value = 0;
        this.onTouchedCallback();
        this.onChangeCallback(this.value);
    };
    TerraCategoryPickerComponent.prototype.updateCompleteCategory = function (category) {
        this.completeCategory.id = +category.id;
        this.completeCategory.isActive = category.isActive;
        this.completeCategory.isOpen = category.isOpen;
        this.completeCategory.isVisible = category.isVisible;
        this.completeCategory.name = category.name;
        this.completeCategory.tooltip = category.tooltip;
        this.completeCategory.tooltipPlacement = category.tooltipPlacement;
    };
    TerraCategoryPickerComponent.prototype.getCategoriesByParentId = function (parentId) {
        var _this = this;
        return function () { return _this.getCategories(parentId); };
    };
    TerraCategoryPickerComponent.prototype.getCategories = function (parentId) {
        var _this = this;
        var obs = this.inputCategoryService.requestCategoryData(parentId);
        obs.subscribe(function (data) {
            _this.addNodes(data, parentId);
        });
        return obs;
    };
    TerraCategoryPickerComponent.prototype.addNodes = function (data, parentNodeId) {
        var _this = this;
        // List of Categories which will be turned into Nodes to add to the node tree
        var entries = data.entries;
        // Necessary for re-initializing of the Node Tree after data was loaded
        if (this.nestedTreeConfig.list.length === 1 && this.nestedTreeConfig.list[0] === this.nestedTreeConfig.currentSelectedNode) {
            this.nestedTreeConfig.removeNodeById(this.nestedTreeConfig.currentSelectedNode.id);
            this.nestedTreeConfig.list = [];
        }
        if (!util_1.isNullOrUndefined(entries)) {
            entries.forEach(function (entry) {
                var categoryData = entry;
                var categoryDetail = null;
                // If the node hasn't already been added the routine will be started
                if (util_1.isNullOrUndefined(_this.nestedTreeConfig.findNodeById(categoryData.id)) && categoryData.details.length > 0) {
                    categoryDetail = categoryData.details[0];
                    // Create Node to add to tree later
                    var childNode = {
                        id: categoryData.id,
                        name: categoryDetail.name,
                        isVisible: true,
                        tooltip: 'ID: ' + categoryData.id,
                        tooltipPlacement: 'top',
                        value: {
                            data: categoryData
                        }
                    };
                    var parentNode = void 0;
                    // If the category has a parent, the parent node is created from the parentId in the category data
                    if (!util_1.isNullOrUndefined(categoryData.parentCategoryId)) {
                        parentNode = _this.nestedTreeConfig.findNodeById(categoryData.parentCategoryId);
                    }
                    // If the parentNode is still null it is tried to create the parent node out of the given id
                    if (util_1.isNullOrUndefined(parentNode)) {
                        if (util_1.isNullOrUndefined(parentNodeId)) {
                            parentNode = null;
                        }
                        else {
                            parentNode = _this.nestedTreeConfig.findNodeById(parentNodeId);
                        }
                    }
                    // If the category has children the lazy-loading method will be added to the parent node
                    if (categoryData.hasChildren) {
                        childNode.onLazyLoad = _this.getCategoriesByParentId(childNode.id);
                    }
                    // The finished node is added to the node tree
                    _this.nestedTreeConfig.addNode(childNode, parentNode);
                }
            });
        }
        // Current List is updated
        this.list = this.nestedTreeConfig.list;
    };
    TerraCategoryPickerComponent.prototype.getCategoriesByParent = function (parentNode) {
        var _this = this;
        var id = null;
        if (!util_1.isNullOrUndefined(parentNode)) {
            id = parentNode.id;
        }
        this.inputCategoryService.requestCategoryData(id).subscribe(function (data) {
            if (util_1.isNullOrUndefined(parentNode)) {
                _this.addNodes(data, id);
            }
            else {
                _this.addNodes(data, parentNode.id);
            }
        });
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", terra_category_picker_base_service_1.TerraCategoryPickerBaseService)
    ], TerraCategoryPickerComponent.prototype, "inputCategoryService", void 0);
    TerraCategoryPickerComponent = TerraCategoryPickerComponent_1 = __decorate([
        core_1.Component({
            selector: 'terra-category-picker',
            template: require('./terra-category-picker.component.html'),
            styles: [require('./terra-category-picker.component.scss')],
            providers: [
                {
                    provide: forms_1.NG_VALUE_ACCESSOR,
                    useExisting: core_1.forwardRef(function () { return TerraCategoryPickerComponent_1; }),
                    multi: true
                },
                __1.TerraNodeTreeConfig,
                nested_data_tree_config_1.NestedDataTreeConfig
            ]
        }),
        __metadata("design:paramtypes", [angular_l10n_1.TranslationService,
            nested_data_tree_config_1.NestedDataTreeConfig])
    ], TerraCategoryPickerComponent);
    return TerraCategoryPickerComponent;
    var TerraCategoryPickerComponent_1;
}(terra_nested_data_picker_component_1.TerraNestedDataPickerComponent));
exports.TerraCategoryPickerComponent = TerraCategoryPickerComponent;
