// @plentymarkets/terra-components/app/components/forms/dynamic-form/terra-dynamic-form.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var util_1 = require("util");
var terra_form_field_control_service_1 = require("./service/terra-form-field-control.service");
var terra_dynamic_form_functions_handler_1 = require("./handler/terra-dynamic-form-functions.handler");
var terra_dynamic_form_service_1 = require("./service/terra-dynamic-form.service");
var operators_1 = require("rxjs/operators");
var TerraHtmlMethods;
(function (TerraHtmlMethods) {
    TerraHtmlMethods["GET"] = "get";
    TerraHtmlMethods["POST"] = "post";
    TerraHtmlMethods["PUT"] = "put";
    TerraHtmlMethods["DELETE"] = "delete";
})(TerraHtmlMethods = exports.TerraHtmlMethods || (exports.TerraHtmlMethods = {}));
/**
 * @author mfrank
 */
var TerraDynamicFormComponent = /** @class */ (function () {
    function TerraDynamicFormComponent(formFieldControlService) {
        this.formFieldControlService = formFieldControlService;
        this.inputUsePortlet = true;
        this.inputPortletStyle = 'col-xs-12 col-md-8 col-lg-5';
        this.inputRequestParams = {
            route: '',
            htmlMethod: null,
            params: {}
        };
        this.inputHasNoSaveButton = false;
        this.inputHasNoResetButton = false;
        this.inputIsDisabled = false;
    }
    TerraDynamicFormComponent.prototype.ngOnInit = function () {
        if (util_1.isNullOrUndefined(this.inputFormFields) || util_1.isNullOrUndefined(this.inputFormFunctions)) {
            console.error('inputFormFields and inputFormFunctions must be set.');
        }
        else {
            this.formFieldControlService.createFormGroup(this.inputFormFields);
            this.inputFormFunctions.formFieldControlService = this.formFieldControlService;
            this.registerValueChange();
        }
    };
    TerraDynamicFormComponent.prototype.ngOnChanges = function (changes) {
        if (changes['inputFormFields']) {
            this.formFieldControlService.createFormGroup(this.inputFormFields);
            this.registerValueChange();
        }
    };
    TerraDynamicFormComponent.prototype.validate = function () {
        if (this.formFieldControlService.dynamicFormGroup.valid) {
            this.inputFormFunctions.saveCallback(this.formFieldControlService.dynamicFormGroup.value);
        }
        else {
            this.inputFormFunctions.errorCallback(this.formFieldControlService.dynamicFormGroup, this.formFieldControlService.translationMapping);
        }
    };
    TerraDynamicFormComponent.prototype.onResetClick = function () {
        this.formFieldControlService.resetForm();
    };
    TerraDynamicFormComponent.prototype.registerValueChange = function () {
        var _this = this;
        if (!util_1.isNullOrUndefined(this.inputFormFunctions.onValueChangedCallback)) {
            var stream$ = this.formFieldControlService
                .dynamicFormGroup
                .valueChanges;
            if (this.inputFormFunctions.valueChangeDebounce > 0) {
                stream$ = stream$.pipe(operators_1.debounceTime(this.inputFormFunctions.valueChangeDebounce));
            }
            stream$.subscribe(function (value) {
                _this.inputFormFunctions.onValueChangedCallback(value);
            });
        }
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", terra_dynamic_form_functions_handler_1.TerraDynamicFormFunctionsHandler)
    ], TerraDynamicFormComponent.prototype, "inputFormFunctions", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array)
    ], TerraDynamicFormComponent.prototype, "inputFormFields", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraDynamicFormComponent.prototype, "inputPortletStyle", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraDynamicFormComponent.prototype, "inputRequestParams", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraDynamicFormComponent.prototype, "inputHasNoSaveButton", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraDynamicFormComponent.prototype, "inputHasNoResetButton", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraDynamicFormComponent.prototype, "inputIsDisabled", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraDynamicFormComponent.prototype, "inputUsePortlet", void 0);
    TerraDynamicFormComponent = __decorate([
        core_1.Component({
            selector: 'terra-dynamic-form',
            template: require('./terra-dynamic-form.component.html'),
            styles: [require('./terra-dynamic-form.component.scss')],
            providers: [terra_dynamic_form_service_1.TerraDynamicFormService]
        }),
        __metadata("design:paramtypes", [terra_form_field_control_service_1.TerraFormFieldControlService])
    ], TerraDynamicFormComponent);
    return TerraDynamicFormComponent;
}());
exports.TerraDynamicFormComponent = TerraDynamicFormComponent;
