// @plentymarkets/terra-components/app/components/buttons/button-with-options/example/terra-button-with-options.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraButtonWithOptionsComponentExample = /** @class */ (function () {
    function TerraButtonWithOptionsComponentExample() {
        this.buttonOptionList = [];
        this.simpleTableHeaderList = [];
        this.simpleTableRowList = [];
        this.listEntryCounter = 0;
    }
    TerraButtonWithOptionsComponentExample.prototype.ngOnInit = function () {
        var _this = this;
        this.buttonOptionList.push({
            caption: 'Add new account',
            icon: 'icon-add',
            clickFunction: function () { return _this.addAccount(); }
        });
        this.buttonOptionList.push({
            caption: 'Delete all accounts',
            icon: 'icon-delete',
            clickFunction: function () { return _this.deleteAllAccounts(); }
        });
        this.simpleTableHeaderList.push({
            caption: 'id'
        }, {
            caption: 'username'
        }, {
            caption: 'mail'
        });
        while (this.listEntryCounter < 5) {
            this.addAccount();
        }
    };
    TerraButtonWithOptionsComponentExample.prototype.addAccount = function () {
        var userName = 'user' + Date.now();
        this.simpleTableRowList.push({
            cellList: [
                {
                    caption: this.listEntryCounter++,
                },
                {
                    caption: userName,
                    icon: 'icon-user_my_account'
                },
                {
                    caption: userName + '@plentymarkets.com',
                    icon: 'icon-email_service'
                },
            ]
        });
    };
    TerraButtonWithOptionsComponentExample.prototype.deleteAllAccounts = function () {
        this.simpleTableRowList = [];
        this.listEntryCounter = 0;
    };
    TerraButtonWithOptionsComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-button-with-options-example',
            styles: [require('./terra-button-with-options.component.example.scss')],
            template: require('./terra-button-with-options.component.example.html'),
        })
    ], TerraButtonWithOptionsComponentExample);
    return TerraButtonWithOptionsComponentExample;
}());
exports.TerraButtonWithOptionsComponentExample = TerraButtonWithOptionsComponentExample;
