// @plentymarkets/terra-components/app/components/file-browser/file-list/file-list.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_file_browser_component_1 = require("../terra-file-browser.component");
var terra_file_browser_service_1 = require("../terra-file-browser.service");
var angular_l10n_1 = require("angular-l10n");
var util_1 = require("util");
var terra_base_private_storage_interface_1 = require("../terra-base-private-storage.interface");
var path_helper_1 = require("../../../helpers/path.helper");
var terra_simple_table_component_1 = require("../../tables/simple/terra-simple-table.component");
var fileType_helper_1 = require("../../../helpers/fileType.helper");
var clipboard_helper_1 = require("../../../helpers/clipboard.helper");
var TerraFileListComponent = /** @class */ (function () {
    function TerraFileListComponent(changeDetector, fileBrowserService, translationService, parentFileBrowser) {
        this.changeDetector = changeDetector;
        this.fileBrowserService = fileBrowserService;
        this.translationService = translationService;
        this.parentFileBrowser = parentFileBrowser;
        this.inputStorageServices = null;
        this.translationPrefix = 'terraFileBrowser';
        this.isDragActive = false;
        this.progress = null;
        this.showNewDirectoryPrompt = false;
        this.objectsToDelete = [];
        this.selectedStorageObjects = [];
        this.fileTableHeaderList = [];
        this.fileTableRowList = [];
        this._newDirectoryName = '';
        this.datePipe = new angular_l10n_1.L10nDatePipe();
    }
    Object.defineProperty(TerraFileListComponent.prototype, "activeStorageService", {
        get: function () {
            if (!util_1.isNullOrUndefined(this._activeStorageService)) {
                return this._activeStorageService;
            }
            if (!util_1.isNullOrUndefined(this.inputStorageServices)) {
                return this.inputStorageServices[0];
            }
            return null;
        },
        set: function (service) {
            var _this = this;
            if (service !== this._activeStorageService) {
                if (!util_1.isNullOrUndefined(this.storageSubscription)) {
                    this.storageSubscription.unsubscribe();
                    this.storageSubscription = null;
                }
                if (!util_1.isNullOrUndefined(this.progressSubscription)) {
                    this.progressSubscription.unsubscribe();
                    this.progressSubscription = null;
                }
                this.storageList = null;
                this._currentStorageRoot = null;
                this._activeStorageService = service;
                if (this.imagePreviewObject) {
                    this.imagePreviewObject = null;
                    this.parentFileBrowser.splitConfig.hideImagePreview();
                }
                this.renderFileList();
                this.storageSubscription = this.activeStorageService.getStorageList().subscribe(function (storageList) {
                    _this.storageList = storageList;
                    _this.renderFileList();
                });
                this.progressSubscription = this.activeStorageService.queue.status.subscribe(function (progress) {
                    _this.progress = progress;
                    if (!util_1.isNullOrUndefined(_this.progress)) {
                        if (util_1.isNumber(_this.progress.sizeUploaded)) {
                            _this.progress.sizeUploaded = path_helper_1.PathHelper.sizeString(_this.progress.sizeUploaded);
                        }
                        if (util_1.isNumber(_this.progress.sizeTotal)) {
                            _this.progress.sizeTotal = path_helper_1.PathHelper.sizeString(_this.progress.sizeTotal);
                        }
                    }
                    _this.changeDetector.detectChanges();
                });
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraFileListComponent.prototype, "currentStorageRoot", {
        get: function () {
            if (!util_1.isNullOrUndefined(this._currentStorageRoot)) {
                return this._currentStorageRoot;
            }
            if (!util_1.isNullOrUndefined(this.storageList)) {
                return this.storageList.root;
            }
            return null;
        },
        set: function (storageObject) {
            if ((util_1.isNullOrUndefined(storageObject) || storageObject.isDirectory) && this._currentStorageRoot !== storageObject) {
                if (this.imagePreviewObject && storageObject !== this.imagePreviewObject) {
                    this.imagePreviewObject = null;
                    this.parentFileBrowser.splitConfig.hideImagePreview();
                }
                this._currentStorageRoot = storageObject;
                this.renderFileList();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraFileListComponent.prototype, "parentStorageObjects", {
        get: function () {
            var current = this.currentStorageRoot;
            var parents = [];
            if (!util_1.isNullOrUndefined(this.imagePreviewObject)) {
                parents.push(this.imagePreviewObject);
            }
            while (!util_1.isNullOrUndefined(current)) {
                parents.push(current);
                current = current.parent;
            }
            return parents.reverse();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraFileListComponent.prototype, "newDirectoryName", {
        get: function () {
            return this._newDirectoryName;
        },
        set: function (name) {
            this._newDirectoryName = this.activeStorageService.prepareKey(name, true, true);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraFileListComponent.prototype, "_deleteCount", {
        get: function () {
            if (util_1.isNullOrUndefined(this.objectsToDelete)) {
                return 0;
            }
            return this.objectsToDelete
                .map(function (object) {
                return object.fileCount;
            })
                .reduce(function (sum, current) {
                return sum + current;
            }, 0);
        },
        enumerable: true,
        configurable: true
    });
    TerraFileListComponent.prototype.ngOnInit = function () {
        var _this = this;
        if (!util_1.isNullOrUndefined(this.inputStorageServices) && this.inputStorageServices.length > 0) {
            this.activeStorageService = this.inputStorageServices[0];
        }
        else {
            console.error('At least one instance of TerraBaseStorageService is required');
        }
        this.dragSubscription = this.fileBrowserService.isDragActive.subscribe(function (isDragActive) {
            _this.isDragActive = isDragActive;
            _this.changeDetector.detectChanges();
        });
        this.parentFileBrowser.onSelectedUrlChange.subscribe(function (selectedUrl) {
            if (selectedUrl && _this.storageList) {
                var object_1 = _this.storageList.flatList.find(function (storage) { return storage.publicUrl === selectedUrl; });
                if (!util_1.isNullOrUndefined(object_1)) {
                    _this.currentStorageRoot = object_1.parent;
                    if (fileType_helper_1.FileTypeHelper.isWebImage(object_1.key)) {
                        _this.imagePreviewObject = object_1;
                        _this.parentFileBrowser.splitConfig.showImagePreview(object_1, _this.activeStorageService);
                    }
                    _this.fileTableComponent.inputHighlightedRow = _this.fileTableRowList.find(function (r) { return r.value === object_1; });
                }
            }
        });
    };
    TerraFileListComponent.prototype.ngAfterViewInit = function () {
        this.fileBrowserService.addDropzone(this.fileDropzoneElement.nativeElement);
    };
    TerraFileListComponent.prototype.ngOnChanges = function (changes) {
        if (changes.hasOwnProperty('inputStorageServices') && util_1.isNull(changes['inputStorageServices'].previousValue)) {
            this.activeStorageService = this.inputStorageServices[0];
        }
    };
    TerraFileListComponent.prototype.ngOnDestroy = function () {
        if (!util_1.isNullOrUndefined(this.storageSubscription)) {
            this.storageSubscription.unsubscribe();
        }
        if (!util_1.isNullOrUndefined(this.dragSubscription)) {
            this.dragSubscription.unsubscribe();
        }
        this.fileBrowserService.removeDropzone(this.fileDropzoneElement.nativeElement);
    };
    TerraFileListComponent.prototype.createDirectory = function () {
        var path = path_helper_1.PathHelper.join(this.currentStorageRoot ? this.currentStorageRoot.key : '/', this.newDirectoryName);
        this.showNewDirectoryPrompt = false;
        this._newDirectoryName = null;
        this.activeStorageService.createDirectory(path);
    };
    TerraFileListComponent.prototype.deleteObjects = function () {
        var keyList = [];
        var extractKeys = function (objects) {
            objects.forEach(function (object) {
                keyList.push(object.key);
                if (object.isDirectory) {
                    extractKeys(object.children);
                }
            });
        };
        extractKeys(this.objectsToDelete);
        this.activeStorageService.deleteFiles(keyList);
        this.objectsToDelete = [];
        if (!util_1.isNullOrUndefined(this.imagePreviewObject) && keyList.indexOf(this.imagePreviewObject.key)) {
            this.imagePreviewObject = null;
            this.parentFileBrowser.splitConfig.hideImagePreview();
        }
    };
    TerraFileListComponent.prototype.renderFileList = function () {
        if (!util_1.isNullOrUndefined(this.activeStorageService)) {
            this.createHeaderListDependingOnAccessLevel();
        }
        if (!util_1.isNullOrUndefined(this.currentStorageRoot)) {
            this.fillTableRowList();
        }
        else {
            this.fileTableRowList = [];
        }
        this.selectedStorageObjects = [];
        this.parentFileBrowser.outputSelectedChange.emit(null);
        this.changeDetector.detectChanges();
    };
    TerraFileListComponent.prototype.fillTableRowList = function () {
        var _this = this;
        this.fileTableRowList = this.currentStorageRoot.children.filter(function (storageObject) {
            return storageObject.isFile || _this.parentFileBrowser.inputAllowFolders;
        }).sort(function (objectA, objectB) {
            return objectA.name.localeCompare(objectB.name);
        }).filter(function (storageObject) { return _this.isAllowed(storageObject.key); })
            .map(function (storageObject) {
            return _this.createTableRow(storageObject);
        });
    };
    TerraFileListComponent.prototype.createTableRow = function (storageObject) {
        var cellList = [];
        cellList.push({
            caption: storageObject.name,
            icon: storageObject.icon
        });
        if (!(this.activeStorageService instanceof terra_base_private_storage_interface_1.TerraBasePrivateStorageService)) {
            cellList.push({
                caption: storageObject.isFile ? storageObject.publicUrl : ''
            });
            this.addClipboardButton(storageObject, cellList);
        }
        cellList.push({
            caption: storageObject.isFile ? storageObject.sizeString : ''
        }, {
            caption: storageObject.isFile ? this.datePipe.transform(storageObject.lastModified, this.defaultLocale, 'medium') : ''
        });
        if (this.activeStorageService instanceof terra_base_private_storage_interface_1.TerraBasePrivateStorageService && storageObject.isFile) {
            this.addDownloadButton(storageObject, cellList);
        }
        this.addDeleteButton(storageObject, cellList);
        return {
            cellList: cellList,
            value: storageObject,
            disabled: !this.isAllowed(storageObject.key)
        };
    };
    TerraFileListComponent.prototype.addClipboardButton = function (storageObject, cellList) {
        var clipboardButton = {
            icon: 'icon-copy_clipboard',
            clickFunction: function (event) {
                clipboard_helper_1.ClipboardHelper.copyText(storageObject.publicUrl);
                event.stopPropagation();
            },
            tooltipText: this.translationService.translate(this.translationPrefix + '.copyToClipboard'),
            tooltipPlacement: 'left'
        };
        cellList.push({
            buttonList: storageObject.isFile ? [clipboardButton] : []
        });
    };
    TerraFileListComponent.prototype.addDownloadButton = function (storageObject, cellList) {
        var _this = this;
        cellList.push({
            buttonList: [{
                    icon: 'icon-download',
                    clickFunction: function (event) {
                        _this.activeStorageService.downloadFile(storageObject.key);
                        event.stopPropagation();
                    },
                    tooltipText: this.translationService.translate(this.translationPrefix + '.downloadFile'),
                    tooltipPlacement: 'left'
                }]
        });
    };
    TerraFileListComponent.prototype.addDeleteButton = function (storageObject, cellList) {
        var _this = this;
        cellList.push({
            buttonList: [{
                    icon: 'icon-delete',
                    clickFunction: function (event) {
                        _this.objectsToDelete = [storageObject];
                        event.stopPropagation();
                    },
                    isSecondary: true,
                    tooltipText: storageObject.isFile ?
                        this.translationService.translate(this.translationPrefix + '.deleteFile') :
                        this.translationService.translate(this.translationPrefix + '.deleteFolder'),
                    tooltipPlacement: 'left'
                }]
        });
    };
    TerraFileListComponent.prototype.createHeaderListDependingOnAccessLevel = function () {
        if (this.activeStorageService instanceof terra_base_private_storage_interface_1.TerraBasePrivateStorageService) {
            this.createPrivateHeaderList();
        }
        else {
            this.createPublicHeaderList();
        }
    };
    TerraFileListComponent.prototype.createPublicHeaderList = function () {
        this.fileTableHeaderList = [
            {
                caption: this.translationService.translate(this.translationPrefix + '.fileName'),
                width: '30%'
            },
            {
                caption: this.translationService.translate(this.translationPrefix + '.fileURL'),
                width: '50%'
            },
            {
                caption: '',
                width: '1'
            },
            {
                caption: this.translationService.translate(this.translationPrefix + '.fileSize'),
                width: '7.5%'
            },
            {
                caption: this.translationService.translate(this.translationPrefix + '.lastChange'),
                width: '12.5%'
            },
            {
                caption: '',
                width: '1'
            }
        ];
    };
    TerraFileListComponent.prototype.createPrivateHeaderList = function () {
        this.fileTableHeaderList = [
            {
                caption: this.translationService.translate(this.translationPrefix + '.fileName'),
                width: '80%'
            },
            {
                caption: this.translationService.translate(this.translationPrefix + '.fileSize'),
                width: '7.5%'
            },
            {
                caption: this.translationService.translate(this.translationPrefix + '.lastChange'),
                width: '12.5%'
            },
            {
                caption: '',
                width: '1'
            },
            {
                caption: '',
                width: '1'
            }
        ];
    };
    TerraFileListComponent.prototype.isAllowed = function (filename) {
        if (util_1.isNullOrUndefined(filename)) {
            return false;
        }
        return this.parentFileBrowser.inputAllowedExtensions.length <= 0
            || this.parentFileBrowser.inputAllowedExtensions.indexOf(path_helper_1.PathHelper.extName(filename).toUpperCase()) >= 0
            || path_helper_1.PathHelper.isDirectory(filename);
    };
    TerraFileListComponent.prototype.onActiveRowChange = function (row) {
        var _this = this;
        if (util_1.isNullOrUndefined(this.imagePreviewTimeout)) {
            clearTimeout(this.imagePreviewTimeout);
        }
        var debounceFn = function () {
            var storageObject = row.value;
            if (!util_1.isNullOrUndefined(storageObject) && fileType_helper_1.FileTypeHelper.isWebImage(storageObject.key)) {
                _this.imagePreviewObject = storageObject;
                _this.parentFileBrowser.splitConfig.showImagePreview(storageObject, _this.activeStorageService);
            }
            else {
                _this.imagePreviewObject = null;
                _this.parentFileBrowser.splitConfig.hideImagePreview();
            }
            _this.parentFileBrowser.outputSelectedChange.emit(storageObject);
        };
        this.imagePreviewTimeout = setTimeout(debounceFn.bind(this), 500);
    };
    TerraFileListComponent.prototype.onSelectionChange = function (rows) {
        this.selectedStorageObjects = rows.map(function (row) {
            return row.value;
        });
    };
    TerraFileListComponent.prototype.onFileSelect = function (event) {
        var target = event.target || event.srcElement;
        if (!util_1.isNullOrUndefined(target) && !util_1.isNullOrUndefined(target.files)) {
            this.uploadFiles(target.files);
            // unset value of file input to allow selecting same file again
            event.target.value = '';
        }
    };
    TerraFileListComponent.prototype.onFileDrop = function (event) {
        if (!util_1.isNullOrUndefined(event.dataTransfer.files)) {
            this.uploadFiles(event.dataTransfer.files);
        }
    };
    TerraFileListComponent.prototype.uploadFiles = function (fileList) {
        var uploadPrefix = this.currentStorageRoot ? this.currentStorageRoot.key : '/';
        this.activeStorageService
            .uploadFiles(fileList, uploadPrefix);
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array)
    ], TerraFileListComponent.prototype, "inputStorageServices", void 0);
    __decorate([
        core_1.ViewChild('fileDropzone', { read: core_1.ElementRef }),
        __metadata("design:type", core_1.ElementRef)
    ], TerraFileListComponent.prototype, "fileDropzoneElement", void 0);
    __decorate([
        core_1.ViewChild(terra_simple_table_component_1.TerraSimpleTableComponent),
        __metadata("design:type", terra_simple_table_component_1.TerraSimpleTableComponent)
    ], TerraFileListComponent.prototype, "fileTableComponent", void 0);
    __decorate([
        angular_l10n_1.DefaultLocale(),
        __metadata("design:type", String)
    ], TerraFileListComponent.prototype, "defaultLocale", void 0);
    TerraFileListComponent = __decorate([
        core_1.Component({
            selector: 'terra-file-list',
            template: require('./file-list.component.html'),
            styles: [require('./file-list.component.scss')]
        }),
        __param(3, core_1.Inject(core_1.forwardRef(function () { return terra_file_browser_component_1.TerraFileBrowserComponent; }))),
        __metadata("design:paramtypes", [core_1.ChangeDetectorRef,
            terra_file_browser_service_1.TerraFileBrowserService,
            angular_l10n_1.TranslationService,
            terra_file_browser_component_1.TerraFileBrowserComponent])
    ], TerraFileListComponent);
    return TerraFileListComponent;
}());
exports.TerraFileListComponent = TerraFileListComponent;
