// @plentymarkets/terra-components/app/components/forms/form/form-container/terra-form-container.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_form_scope_data_1 = require("../model/terra-form-scope.data");
var util_1 = require("util");
var TerraFormContainerComponent = /** @class */ (function () {
    function TerraFormContainerComponent() {
        this.inputControlTypeMap = {};
        this.inputIsDisabled = false;
        this.outputFormValueChanged = new core_1.EventEmitter();
        this.formFields = [];
        this.formFieldVisibility = {};
        this.value = {};
    }
    Object.defineProperty(TerraFormContainerComponent.prototype, "inputValue", {
        get: function () {
            return this.value;
        },
        set: function (value) {
            if (util_1.isNullOrUndefined(value)) {
                this.value = {};
            }
            else {
                this.value = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraFormContainerComponent.prototype, "inputFormFields", {
        set: function (fields) {
            this.formFields = Object.keys(fields)
                .map(function (key) {
                return {
                    key: key,
                    value: fields[key]
                };
            });
            this.updateFieldVisibility();
        },
        enumerable: true,
        configurable: true
    });
    TerraFormContainerComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.inputScope.onDataChanged.subscribe(function (data) {
            _this.updateFieldVisibility();
        });
    };
    TerraFormContainerComponent.prototype.ngOnChanges = function (changes) {
        if (changes.hasOwnProperty('inputScope')) {
            this.updateFieldVisibility();
        }
    };
    TerraFormContainerComponent.prototype.onFormValueChanged = function (key, value) {
        this.value[key] = value;
        this.updateFieldVisibility();
        this.outputFormValueChanged.next({
            key: key,
            value: value
        });
    };
    TerraFormContainerComponent.prototype.updateFieldVisibility = function () {
        var _this = this;
        this.formFields
            .forEach(function (field) {
            if (util_1.isString(field.value.isVisible)) {
                _this.formFieldVisibility[field.key] = _this.inputScope.evaluate(field.value.isVisible);
            }
            else {
                _this.formFieldVisibility[field.key] = util_1.isNullOrUndefined(field.value.isVisible) || field.value.isVisible;
            }
        });
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", terra_form_scope_data_1.TerraFormScope)
    ], TerraFormContainerComponent.prototype, "inputScope", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraFormContainerComponent.prototype, "inputControlTypeMap", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], TerraFormContainerComponent.prototype, "inputValue", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], TerraFormContainerComponent.prototype, "inputFormFields", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraFormContainerComponent.prototype, "inputIsDisabled", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraFormContainerComponent.prototype, "outputFormValueChanged", void 0);
    TerraFormContainerComponent = __decorate([
        core_1.Component({
            selector: 'terra-form-container',
            template: require('./terra-form-container.component.html'),
            styles: [require('./terra-form-container.component.scss')]
        })
    ], TerraFormContainerComponent);
    return TerraFormContainerComponent;
}());
exports.TerraFormContainerComponent = TerraFormContainerComponent;
