// @plentymarkets/terra-components/app/components/tables/fixed-header/fixed-header.directive.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_simple_table_component_1 = require("../../tables/simple/terra-simple-table.component");
var FIXED_CLASS = 'fixedHeader';
var FixedHeaderDirective = /** @class */ (function () {
    function FixedHeaderDirective(elementRef, tableComponent) {
        this.elementRef = elementRef;
        this.tableComponent = tableComponent;
        this.columnWidths = [];
    }
    FixedHeaderDirective.prototype.ngAfterViewInit = function () {
        this.tableElement = this.elementRef.nativeElement.querySelector('table');
        this.tableHeadElement = this.tableElement.querySelector('thead');
        this.tableBodyElement = this.tableElement.querySelector('tbody');
    };
    FixedHeaderDirective.prototype.ngAfterViewChecked = function () {
        // check if table has at least one row
        if (this.tableElement && this.tableBodyElement.querySelector('tr:first-child')) {
            this.updateColumnWidths();
        }
    };
    FixedHeaderDirective.prototype.updateColumnWidths = function () {
        var _this = this;
        var rows = this.tableBodyElement.querySelectorAll('tr');
        var headerCol;
        var bodyCol;
        // adjust difference between header width and body width
        var headerWidth = this.tableHeadElement.getBoundingClientRect().width;
        var bodyWidth = this.tableBodyElement.getBoundingClientRect().width;
        this.tableHeadElement.style.paddingRight = (headerWidth - bodyWidth) + 'px';
        // assign column widths
        this.getColumnWidths()
            .forEach(function (width, index) {
            headerCol = _this.tableHeadElement.querySelector('tr th:nth-child(' + (index + 1) + ')');
            if (headerCol) {
                headerCol.style.width = width + '%';
            }
            for (var i = 0; i < rows.length; i++) {
                bodyCol = rows.item(i).querySelector('tr td:nth-child(' + (index + 1) + ')');
                if (bodyCol) {
                    bodyCol.style.width = width + '%';
                }
            }
        });
    };
    FixedHeaderDirective.prototype.getColumnWidths = function () {
        var firstRow = this.tableBodyElement.querySelector('tr:first-child');
        if (firstRow) {
            var rowWidth = firstRow.getBoundingClientRect().width;
            var columns = firstRow.querySelectorAll('td');
            if (rowWidth > 0 && columns.length !== this.columnWidths.length) {
                this.columnWidths = [];
                for (var i = 0; i < columns.length; i++) {
                    this.columnWidths.push((columns.item(i).getBoundingClientRect().width / rowWidth) * 100);
                }
            }
        }
        return this.columnWidths;
    };
    FixedHeaderDirective = __decorate([
        core_1.Directive({
            selector: 'terra-simple-table[fixedHeader]'
        }),
        __param(1, core_1.Inject(core_1.forwardRef(function () { return terra_simple_table_component_1.TerraSimpleTableComponent; }))),
        __metadata("design:paramtypes", [core_1.ElementRef, Object])
    ], FixedHeaderDirective);
    return FixedHeaderDirective;
}());
exports.FixedHeaderDirective = FixedHeaderDirective;
