// @plentymarkets/terra-components/app/components/navigator/button-group/terra-button-group.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_navigator_split_view_config_1 = require("../config/terra-navigator-split-view.config");
var util_1 = require("util");
/**
 * @author mscharf
 * @deprecated since `terra-navigator` is now deprecated
 */
var TerraButtonGroupComponent = /** @class */ (function () {
    function TerraButtonGroupComponent(terraNavigatorSplitViewConfig) {
        this.terraNavigatorSplitViewConfig = terraNavigatorSplitViewConfig;
        this.buttonList = [];
    }
    TerraButtonGroupComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.terraNavigatorSplitViewConfig.observableUpdateActiveItem
            .subscribe(function (nodes) {
            _this.setRecursiveItemActive(nodes);
        });
        this.parameter.nodes
            .forEach(function (item) {
            var hasChildren = false;
            if (!util_1.isNullOrUndefined(item.children)) {
                hasChildren = true;
            }
            var button = {
                caption: item.nodeName,
                clickFunction: function () {
                    _this.terraNavigatorSplitViewConfig.openNextLevel(item);
                    item.isButtonClicked = true;
                    _this.buttonList
                        .forEach(function (btnItem) {
                        btnItem.isActive = false;
                    });
                    button.isActive = true;
                },
                hasChildren: hasChildren,
                isVisible: util_1.isNullOrUndefined(item.isVisible) || item.isVisible,
            };
            _this.buttonList
                .push(button);
            if (!util_1.isNullOrUndefined(item.nodeIcon)) {
                _this.buttonList[_this.buttonList.length - 1].icon = item.nodeIcon;
            }
        });
        this.setRecursiveItemActive(this.parameter.nodes);
    };
    TerraButtonGroupComponent.prototype.setRecursiveItemActive = function (list) {
        var _this = this;
        list.forEach(function (item) {
            if (item.isActive) {
                setTimeout(function () {
                    _this.terraNavigatorSplitViewConfig
                        .openNextLevel(item);
                });
                var btn = _this.buttonList.find(function (button) {
                    return button.caption === item.nodeName;
                });
                if (!util_1.isNullOrUndefined(btn)) {
                    btn.isActive = item.isActive;
                }
            }
            else {
                if (item.children) {
                    _this.setRecursiveItemActive(item.children);
                }
            }
        });
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], TerraButtonGroupComponent.prototype, "parameter", void 0);
    TerraButtonGroupComponent = __decorate([
        core_1.Component({
            selector: 'terra-button-group',
            template: require('./terra-button-group.component.html'),
            styles: [require('./terra-button-group.component.scss')]
        }),
        __metadata("design:paramtypes", [terra_navigator_split_view_config_1.TerraNavigatorSplitViewConfig])
    ], TerraButtonGroupComponent);
    return TerraButtonGroupComponent;
}());
exports.TerraButtonGroupComponent = TerraButtonGroupComponent;
