// @plentymarkets/terra-components/app/components/editors/syntax-editor/example/terra-syntax-editor.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraSyntaxEditorComponentExample = /** @class */ (function () {
    function TerraSyntaxEditorComponentExample() {
    }
    TerraSyntaxEditorComponentExample.prototype.ngOnInit = function () {
        this.inputText = "<!DOCTYPE html>\n<html>\n<head>\n    <meta CHARSET=\"UTF-8\">\n    <title>Website</title>\n</head>\n<body>\n    <h1>Hello</h1>\n</body>\n</html>";
        this.inputOptions = {
            maxLines: 10,
            inputText: this.inputText
        };
    };
    TerraSyntaxEditorComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-syntax-editor-example',
            styles: [require('./terra-syntax-editor.component.example.scss')],
            template: require('./terra-syntax-editor.component.example.html'),
        })
    ], TerraSyntaxEditorComponentExample);
    return TerraSyntaxEditorComponentExample;
}());
exports.TerraSyntaxEditorComponentExample = TerraSyntaxEditorComponentExample;
