// @plentymarkets/terra-components/app/components/forms/dynamic-form/enum/terra-control-type.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @author mfrank
 */
var TerraControlTypeEnum;
(function (TerraControlTypeEnum) {
    TerraControlTypeEnum["CHECK_BOX"] = "checkBox";
    TerraControlTypeEnum["CONDITIONAL_CONTAINER"] = "conditionalContainer";
    TerraControlTypeEnum["DATE_PICKER"] = "datePicker";
    TerraControlTypeEnum["HORIZONTAL_CONTAINER"] = "horizontalContainer";
    TerraControlTypeEnum["INPUT_FILE"] = "inputFile";
    TerraControlTypeEnum["INPUT_TEXT"] = "inputText";
    TerraControlTypeEnum["INPUT_TEXT_AREA"] = "inputTextArea";
    TerraControlTypeEnum["INPUT_NUMBER"] = "inputNumber";
    TerraControlTypeEnum["INPUT_DOUBLE"] = "inputDouble";
    TerraControlTypeEnum["SELECT_BOX"] = "selectBox";
    TerraControlTypeEnum["VERTICAL_CONTAINER"] = "verticalContainer";
    TerraControlTypeEnum["CATEGORY_PICKER"] = "categoryPicker";
    TerraControlTypeEnum["COLOR_PICKER"] = "colorPicker";
    TerraControlTypeEnum["MULTI_CHECK_BOX"] = "multiCheckBox";
    TerraControlTypeEnum["NOTE_EDITOR"] = "noteEditor";
    TerraControlTypeEnum["CODE_EDITOR"] = "codeEditor";
})(TerraControlTypeEnum = exports.TerraControlTypeEnum || (exports.TerraControlTypeEnum = {}));
