// @plentymarkets/terra-components/app/components/forms/input/text-area-input/terra-text-area-input.component.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_input_component_1 = require("../terra-input.component");
var forms_1 = require("@angular/forms");
var terra_regex_1 = require("../../../../helpers/regex/terra-regex");
var util_1 = require("util");
var nextId = 0;
var TerraTextAreaInputComponent = /** @class */ (function (_super) {
    __extends(TerraTextAreaInputComponent, _super);
    function TerraTextAreaInputComponent() {
        var _this = _super.call(this, terra_regex_1.TerraRegex.MIXED) || this;
        _this.defaultMaxRows = 4;
        // generate the id of the input instance
        _this.id = "text-area-input_#" + nextId++;
        _this.inputMaxRows = _this.defaultMaxRows;
        _this.inputHasFixedHeight = false;
        return _this;
    }
    TerraTextAreaInputComponent_1 = TerraTextAreaInputComponent;
    Object.defineProperty(TerraTextAreaInputComponent.prototype, "inputType", {
        /**
         * @deprecated inputType is no longer used.  It will be removed in one of the upcoming releases.
         * @param v
         */
        set: function (v) {
            console.warn('inputType is no longer used.  It will be removed in one of the upcoming releases.');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraTextAreaInputComponent.prototype, "inputValue", {
        /**
         * @deprecated inputValue is deprecated. It will be removed in one of the upcoming releases. Please use ngModel instead.
         * @param v
         */
        set: function (v) {
            console.warn('inputValue is deprecated. It will be removed in one of the upcoming releases. Please use ngModel instead.');
            this.value = v;
        },
        enumerable: true,
        configurable: true
    });
    TerraTextAreaInputComponent.prototype.ngOnChanges = function (changes) {
        if (changes.hasOwnProperty('inputMaxRows') && !util_1.isNullOrUndefined(changes.inputMaxRows)) {
            this.inputMaxRows = Math.max(this.defaultMaxRows, changes.inputMaxRows.currentValue);
        }
        if (changes.hasOwnProperty('inputHasFixedHeight')) {
            this.inputHasFixedHeight = !!changes.inputHasFixedHeight.currentValue;
        }
    };
    /**
     * Set the focus on the native input element.
     */
    TerraTextAreaInputComponent.prototype.focusNativeInput = function () {
        var _this = this;
        setTimeout(function () {
            var input = document.getElementById(_this.id);
            input.focus();
        });
    };
    /**
     * Select the content of the native input element.
     */
    TerraTextAreaInputComponent.prototype.selectNativeInput = function () {
        var _this = this;
        setTimeout(function () {
            var input = document.getElementById(_this.id);
            input.select();
        });
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraTextAreaInputComponent.prototype, "inputIsRequired", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], TerraTextAreaInputComponent.prototype, "inputHasFixedHeight", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], TerraTextAreaInputComponent.prototype, "inputType", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], TerraTextAreaInputComponent.prototype, "inputValue", null);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number)
    ], TerraTextAreaInputComponent.prototype, "inputMaxRows", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number)
    ], TerraTextAreaInputComponent.prototype, "inputMaxCols", void 0);
    TerraTextAreaInputComponent = TerraTextAreaInputComponent_1 = __decorate([
        core_1.Component({
            selector: 'terra-text-area-input',
            styles: [require('./terra-text-area-input.component.scss')],
            template: require('./terra-text-area-input.component.html'),
            providers: [
                {
                    provide: forms_1.NG_VALUE_ACCESSOR,
                    useExisting: core_1.forwardRef(function () { return TerraTextAreaInputComponent_1; }),
                    multi: true
                }
            ]
        }),
        __metadata("design:paramtypes", [])
    ], TerraTextAreaInputComponent);
    return TerraTextAreaInputComponent;
    var TerraTextAreaInputComponent_1;
}(terra_input_component_1.TerraInputComponent));
exports.TerraTextAreaInputComponent = TerraTextAreaInputComponent;
