// @plentymarkets/terra-components/app/components/layouts/portlet/example/terra-portlet.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var TerraPortletComponentExample = /** @class */ (function () {
    function TerraPortletComponentExample() {
        this.portletButtonList = [];
    }
    TerraPortletComponentExample_1 = TerraPortletComponentExample;
    TerraPortletComponentExample.buttonClickFunction = function () {
        alert('button clicked');
    };
    TerraPortletComponentExample.prototype.ngOnInit = function () {
        this.portletButtonList.push({
            icon: 'icon-delete',
            isSecondary: true,
            clickFunction: function () { return TerraPortletComponentExample_1.buttonClickFunction(); }
        }, {
            icon: 'icon-edit',
            clickFunction: function () { return TerraPortletComponentExample_1.buttonClickFunction(); }
        });
    };
    TerraPortletComponentExample = TerraPortletComponentExample_1 = __decorate([
        core_1.Component({
            selector: 'terra-portlet-example',
            template: require('./terra-portlet.component.example.html')
        })
    ], TerraPortletComponentExample);
    return TerraPortletComponentExample;
    var TerraPortletComponentExample_1;
}());
exports.TerraPortletComponentExample = TerraPortletComponentExample;
