// @plentymarkets/terra-components/app/components/editors/code-editor/helper/html-linter-rule.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var HtmlLinterRule;
(function (HtmlLinterRule) {
    HtmlLinterRule["tagnameLowercase"] = "tagname-lowercase";
    HtmlLinterRule["attrLowercase"] = "attr-lowercase";
    HtmlLinterRule["attrValueDoubleQuotes"] = "attr-value-double-quotes";
    HtmlLinterRule["attrValueNotEmpty"] = "attr-value-not-empty";
    HtmlLinterRule["attrNoDuplicate"] = "attr-no-duplication";
    HtmlLinterRule["doctypeFirst"] = "doctype-first";
    HtmlLinterRule["tagPair"] = "tag-pair";
    HtmlLinterRule["tagSelfClose"] = "tag-self-close";
    HtmlLinterRule["specCharEscape"] = "spec-char-escape";
    HtmlLinterRule["idUnique"] = "id-unique";
    HtmlLinterRule["srcNotEmpty"] = "src-not-empty";
    HtmlLinterRule["titleRequire"] = "title-require";
    HtmlLinterRule["altRequire"] = "alt-require";
    HtmlLinterRule["doctypeHtml5"] = "doctype-html5";
    HtmlLinterRule["styleDisabled"] = "style-disabled";
    HtmlLinterRule["inlineStyleDisabled"] = "inline-style-disabled";
    HtmlLinterRule["inlineScriptDisabled"] = "inline-script-disabled";
    HtmlLinterRule["hrefAbsOrRel"] = "href-abs-or-rel";
    HtmlLinterRule["attrUnsafeChars"] = "attr-unsafe-chars";
    HtmlLinterRule["headScriptDisabled"] = "head-script-disabled";
})(HtmlLinterRule = exports.HtmlLinterRule || (exports.HtmlLinterRule = {}));
