// @plentymarkets/terra-components/app/components/dynamic-component-loader/terra-dynamic-component-loader.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var util_1 = require("util");
var TerraDynamicComponentLoaderComponent = /** @class */ (function () {
    function TerraDynamicComponentLoaderComponent(componentFactoryResolver) {
        this.componentFactoryResolver = componentFactoryResolver;
    }
    TerraDynamicComponentLoaderComponent.prototype.ngOnChanges = function (changes) {
        if (changes['inputComponent']) {
            this.ngOnDestroy();
            this.updateComponent();
        }
    };
    TerraDynamicComponentLoaderComponent.prototype.ngAfterViewInit = function () {
        this.updateComponent();
    };
    TerraDynamicComponentLoaderComponent.prototype.ngOnDestroy = function () {
        if (this.cmpRef) {
            this.cmpRef.destroy();
        }
    };
    TerraDynamicComponentLoaderComponent.prototype.updateComponent = function () {
        if (!util_1.isNullOrUndefined(this.inputComponent)) {
            var componentFactory = this.componentFactoryResolver.resolveComponentFactory(this.inputComponent);
            var viewContainerRef = this.viewChildTarget;
            viewContainerRef.clear();
            this.cmpRef = viewContainerRef.createComponent(componentFactory);
        }
    };
    __decorate([
        core_1.ViewChild('viewChildTarget', { read: core_1.ViewContainerRef }),
        __metadata("design:type", core_1.ViewContainerRef)
    ], TerraDynamicComponentLoaderComponent.prototype, "viewChildTarget", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", core_1.Type)
    ], TerraDynamicComponentLoaderComponent.prototype, "inputComponent", void 0);
    TerraDynamicComponentLoaderComponent = __decorate([
        core_1.Component({
            selector: 'terra-dynamic-component-loader',
            template: require('./terra-dynamic-component-loader.component.html'),
        }),
        __metadata("design:paramtypes", [core_1.ComponentFactoryResolver])
    ], TerraDynamicComponentLoaderComponent);
    return TerraDynamicComponentLoaderComponent;
}());
exports.TerraDynamicComponentLoaderComponent = TerraDynamicComponentLoaderComponent;
