// @plentymarkets/terra-components/app/components/alert/example/terra-alert.component.example.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_alert_component_1 = require("../terra-alert.component");
var angular_l10n_1 = require("angular-l10n");
var TerraAlertComponentExample = /** @class */ (function () {
    function TerraAlertComponentExample(translation) {
        this.translation = translation;
        this.exampleAlert = terra_alert_component_1.TerraAlertComponent.getInstance();
    }
    TerraAlertComponentExample.prototype.ngOnInit = function () {
        this.exampleAlert.closeAlertByIdentifier('info');
    };
    TerraAlertComponentExample.prototype.showInformationAlert = function () {
        this.exampleAlert.addAlert({
            msg: 'info-Alert',
            type: 'info',
            dismissOnTimeout: 5000,
            identifier: 'info'
        });
        this.emptyAlertArray();
    };
    TerraAlertComponentExample.prototype.showSuccessAlert = function () {
        this.exampleAlert.addAlert({
            msg: 'success-Alert',
            type: 'success',
            dismissOnTimeout: 5000,
            identifier: 'info'
        });
        this.emptyAlertArray();
    };
    TerraAlertComponentExample.prototype.showErrorAlert = function () {
        this.exampleAlert.addAlert({
            msg: 'error-Alert',
            type: 'danger',
            dismissOnTimeout: 5000,
            identifier: 'info'
        });
        this.emptyAlertArray();
    };
    TerraAlertComponentExample.prototype.showWarningAlert = function () {
        this.exampleAlert.addAlert({
            msg: 'warning-Alert',
            type: 'warning',
            dismissOnTimeout: 5000,
            identifier: 'info'
        });
        this.emptyAlertArray();
    };
    TerraAlertComponentExample.prototype.emptyAlertArray = function () {
        var _this = this;
        setTimeout(function () { return _this.exampleAlert.closeAlertByIdentifier('info'); }, 5000);
    };
    TerraAlertComponentExample = __decorate([
        core_1.Component({
            selector: 'terra-alert-example',
            styles: [require('./terra-alert.component.example.scss')],
            template: require('./terra-alert.component.example.html')
        }),
        __metadata("design:paramtypes", [angular_l10n_1.TranslationService])
    ], TerraAlertComponentExample);
    return TerraAlertComponentExample;
}());
exports.TerraAlertComponentExample = TerraAlertComponentExample;
