// @plentymarkets/terra-components/app/components/editors/base-editor/terra-base-editor.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var angular_l10n_1 = require("angular-l10n");
var forms_1 = require("@angular/forms");
var util_1 = require("util");
var TerraBaseEditorComponent = /** @class */ (function () {
    function TerraBaseEditorComponent(translation, myElement) {
        this.translation = translation;
        this.myElement = myElement;
        this.onChangeCallback = function () { return undefined; };
        this.onTouchedCallback = function () { return undefined; };
        // initialize placeholder
        this.placeholder = this.translation.translate('terraNoteEditor.insertText');
        this.modules = {
            toolbar: [
                ['bold',
                    'italic',
                    'underline',
                    'strike'] // toggled buttons
            ]
        };
    }
    TerraBaseEditorComponent_1 = TerraBaseEditorComponent;
    TerraBaseEditorComponent.prototype.ngOnInit = function () {
        this.inputMinHeight = util_1.isNullOrUndefined(this.inputMinHeight) ? '100px' : this.inputMinHeight;
        // overwrite default placeholder if input is defined
        if (this.inputPlaceholder) {
            this.placeholder = this.inputPlaceholder;
        }
    };
    TerraBaseEditorComponent.prototype.writeValue = function (value) {
        this.value = value;
    };
    TerraBaseEditorComponent.prototype.registerOnChange = function (fn) {
        this.onChangeCallback = fn;
    };
    TerraBaseEditorComponent.prototype.registerOnTouched = function (fn) {
        this.onTouchedCallback = fn;
    };
    TerraBaseEditorComponent.prototype.focus = function () {
        this.myElement.nativeElement.querySelector('.ql-editor').focus();
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraBaseEditorComponent.prototype, "inputHeaderLabel", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraBaseEditorComponent.prototype, "inputPlaceholder", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraBaseEditorComponent.prototype, "inputFixedHeight", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraBaseEditorComponent.prototype, "inputMinHeight", void 0);
    TerraBaseEditorComponent = TerraBaseEditorComponent_1 = __decorate([
        core_1.Component({
            selector: 'terra-base-editor',
            template: require('./terra-base-editor.component.html'),
            styles: [
                require('./terra-base-editor.component.scss'),
                require('quill/dist/quill.bubble.css'),
                require('quill/dist/quill.snow.css'),
                require('./terra-base-editor.component.glob.scss').toString()
            ],
            providers: [{
                    provide: forms_1.NG_VALUE_ACCESSOR,
                    useExisting: TerraBaseEditorComponent_1,
                    multi: true
                }]
        }),
        __metadata("design:paramtypes", [angular_l10n_1.TranslationService,
            core_1.ElementRef])
    ], TerraBaseEditorComponent);
    return TerraBaseEditorComponent;
    var TerraBaseEditorComponent_1;
}());
exports.TerraBaseEditorComponent = TerraBaseEditorComponent;
