// @plentymarkets/terra-components/app/helpers/clipboard.helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ClipboardHelper = /** @class */ (function () {
    function ClipboardHelper() {
    }
    ClipboardHelper.copySelection = function () {
        var success = false;
        try {
            success = document.execCommand('copy');
        }
        catch (err) {
            success = false;
        }
        return success;
    };
    ClipboardHelper.copyText = function (text) {
        var container = ClipboardHelper.createElement('div', {
            position: 'fixed',
            left: '0px',
            top: '0px',
            width: '0px',
            height: '0px',
            display: 'block',
            opacity: '0',
            zIndex: '10000'
        });
        var textarea = ClipboardHelper.createElement('textarea', {
            width: '1px',
            height: '1px',
            padding: '0px'
        });
        textarea.value = text;
        container.appendChild(textarea);
        document.body.appendChild(container);
        textarea.focus();
        textarea.select();
        var success = ClipboardHelper.copySelection();
        document.body.removeChild(container);
        return success;
    };
    ClipboardHelper.createElement = function (tagName, styles) {
        if (styles === void 0) { styles = {}; }
        var element = document.createElement(tagName);
        Object.keys(styles).forEach(function (style) {
            element.style[style] = styles[style];
        });
        return element;
    };
    return ClipboardHelper;
}());
exports.ClipboardHelper = ClipboardHelper;
