// @plentymarkets/terra-components/app/components/filter/terra-filter.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
/**
 * @author mscharf
 */
var TerraFilterComponent = /** @class */ (function () {
    function TerraFilterComponent() {
        /**
         * @description Set the function which will be executed on click of the search button.
         */
        this.outputOnSearchBtnClicked = new core_1.EventEmitter();
        /**
         * @description Set the function which will be executed on click of the reset button.
         */
        this.outputOnResetBtnClicked = new core_1.EventEmitter();
        /**
         * @description Set the function which will be executed on hitting enter.
         */
        this.outputOnEnterSubmit = new core_1.EventEmitter();
    }
    TerraFilterComponent.prototype.searchBtnClicked = function () {
        this.outputOnSearchBtnClicked.emit(null);
    };
    TerraFilterComponent.prototype.resetBtnClicked = function () {
        this.outputOnResetBtnClicked.emit(null);
    };
    TerraFilterComponent.prototype.onSubmit = function () {
        this.outputOnEnterSubmit.emit(null);
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraFilterComponent.prototype, "inputSearchLabel", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], TerraFilterComponent.prototype, "inputResetLabel", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Array)
    ], TerraFilterComponent.prototype, "inputInputList", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraFilterComponent.prototype, "outputOnSearchBtnClicked", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraFilterComponent.prototype, "outputOnResetBtnClicked", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", core_1.EventEmitter)
    ], TerraFilterComponent.prototype, "outputOnEnterSubmit", void 0);
    TerraFilterComponent = __decorate([
        core_1.Component({
            selector: 'terra-filter',
            template: require('./terra-filter.component.html')
        })
    ], TerraFilterComponent);
    return TerraFilterComponent;
}());
exports.TerraFilterComponent = TerraFilterComponent;
