// @plentymarkets/terra-components/app/components/file-browser/image-preview/image-preview.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var terra_storage_object_1 = require("../model/terra-storage-object");
var terra_base_storage_interface_1 = require("../terra-base-storage.interface");
var util_1 = require("util");
var terra_base_metadata_storage_interface_1 = require("../terra-base-metadata-storage.interface");
var TerraImagePreviewComponent = /** @class */ (function () {
    function TerraImagePreviewComponent(changeDetector) {
        this.changeDetector = changeDetector;
        this.translationPrefix = 'terraFileBrowser';
        this.metadata = {};
        this.isLoading = true;
    }
    Object.defineProperty(TerraImagePreviewComponent.prototype, "inputStorageObject", {
        get: function () {
            return this._inputStorageObject;
        },
        set: function (object) {
            var _this = this;
            this._inputStorageObject = object;
            this.metadata = {};
            this.isLoading = true;
            if (!util_1.isNullOrUndefined(object) && this.inputStorageService && this.inputStorageService instanceof terra_base_metadata_storage_interface_1.TerraBaseMetadataStorageService) {
                this.inputStorageService.getMetadata(object.key).subscribe(function (data) {
                    _this.metadata = data;
                    _this.isLoading = false;
                    _this.changeDetector.detectChanges();
                });
            }
            else {
                this.isLoading = false;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TerraImagePreviewComponent.prototype, "_canHandleMetadata", {
        get: function () {
            return this.inputStorageService instanceof terra_base_metadata_storage_interface_1.TerraBaseMetadataStorageService;
        },
        enumerable: true,
        configurable: true
    });
    TerraImagePreviewComponent.prototype.updateMetadata = function () {
        if (this.inputStorageService instanceof terra_base_metadata_storage_interface_1.TerraBaseMetadataStorageService) {
            this.inputStorageService.updateMetadata(this.inputStorageObject.key, this.metadata);
        }
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", terra_base_storage_interface_1.TerraBaseStorageService)
    ], TerraImagePreviewComponent.prototype, "inputStorageService", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", terra_storage_object_1.TerraStorageObject),
        __metadata("design:paramtypes", [terra_storage_object_1.TerraStorageObject])
    ], TerraImagePreviewComponent.prototype, "inputStorageObject", null);
    TerraImagePreviewComponent = __decorate([
        core_1.Component({
            selector: 'terra-image-preview',
            template: require('./image-preview.component.html'),
            styles: [require('./image-preview.component.scss')]
        }),
        __metadata("design:paramtypes", [core_1.ChangeDetectorRef])
    ], TerraImagePreviewComponent);
    return TerraImagePreviewComponent;
}());
exports.TerraImagePreviewComponent = TerraImagePreviewComponent;
